{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 399,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os, sys,time\n",
    "import vcf\n",
    "import gzip\n",
    "from Bio import SeqIO\n",
    "from collections import OrderedDict, Counter\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcftools = \"/home/lindb/g/src/vcftools_0.1.13/bin/vcftools\"\n",
    "bcftools = \"/home/lindb/g/src/bcftools-1.3/bcftools\"\n",
    "tabix = \"/home/lindb/g/src/htslib-1.3/tabix\"\n",
    "bgzip = \"/home/lindb/g/src/htslib-1.3/bgzip\"\n",
    "plink = \"/home/lindb/g/src/plink-1.07-x86_64/plink --noweb\"\n",
    "java  = \"/home/lindb/g/src/jre1.8.0_73/bin/java\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "snp_dir = '/home/lindb/wbp/concatenated/snps/'\n",
    "vcf_file = os.path.join(snp_dir,'samtools_1.3.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "assert os.path.exists(vcf_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--max-alleles 2\n",
      "\t--min-alleles 2\n",
      "\t--max-missing 0.5\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 1300961 out of a possible 2892582 Sites\n",
      "Run Time = 5448.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --remove-indels \\\n",
    "--max-missing 0.5 \\\n",
    "--min-alleles 2 \\\n",
    "--max-alleles 2 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_file \\\n",
    "--out $vcf_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcf_filtered = \"%s.recode.vcf\" % vcf_file\n",
    "vcf_filtered_gz = \"%s.gz\" % vcf_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n"
     ]
    }
   ],
   "source": [
    "!$bgzip -c $vcf_filtered > {vcf_filtered_gz}\n",
    "!$tabix {vcf_filtered_gz}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_vcf_stats(vcf_gz):\n",
    "    \n",
    "    stats = ['depth',\n",
    "            'site-depth',\n",
    "            'site-mean-depth',\n",
    "            'site-quality',\n",
    "            'missing-indv',\n",
    "            'missing-site',\n",
    "            'freq',\n",
    "            'counts',\n",
    "            'hardy',\n",
    "            'het']\n",
    "    \n",
    "    for stat in stats:\n",
    "        !$vcftools --gzvcf $vcf_gz \\\n",
    "        --out $vcf_gz \\\n",
    "        {\"--%s\" % stat} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--depth\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 1207.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--site-depth\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 1199.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--site-mean-depth\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 273.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--site-quality\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Quality for Each Site\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 151.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--missing-indv\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Individual Missingness\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 248.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--missing-site\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Site Missingness\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 258.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--freq\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 251.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--counts\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 249.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--hardy\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting HWE statistics (but only for biallelic loci)\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 282.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--het\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Individual Heterozygosity\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 255.00 seconds\n"
     ]
    }
   ],
   "source": [
    "get_vcf_stats(vcf_filtered_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "files = os.listdir(snp_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz.ldepth',\n",
       " '/home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz.ldepth.mean',\n",
       " '/home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz.lqual',\n",
       " '/home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz.lmiss']"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[os.path.join(snp_dir,f) for f in files if '.l' in f and 'log' not in f]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', 100)\n",
    "\n",
    "def get_MAF(row):\n",
    "    try:\n",
    "        return np.min([row.A1_freq, row.A2_freq])\n",
    "    except:\n",
    "        print(row)\n",
    "        \n",
    "def get_correction(n):\n",
    "    #for finite sample size\n",
    "    return (2*n)/(2*n-1)\n",
    "\n",
    "def calculate_Fis(vals):\n",
    "    try:\n",
    "        data = [float(x) for x in vals.split(\"/\")]\n",
    "        assert len(data) == 3\n",
    "        num_individuals = np.sum(data)\n",
    "        total_alleles = 2*num_individuals\n",
    "        a1_count = 2*data[0]\n",
    "        a2_count = 2*data[2]\n",
    "        het_count = data[1]\n",
    "        a1_count += het_count\n",
    "        a2_count += het_count\n",
    "        a1_freq = a1_count/total_alleles\n",
    "        a2_freq = a2_count/total_alleles\n",
    "        assert a1_freq + a2_freq == 1.0\n",
    "        He = 2 * a1_freq * a2_freq * get_correction(num_individuals)\n",
    "        Ho = het_count/num_individuals\n",
    "        Fis = 1 - (Ho/He)\n",
    "        return Fis\n",
    "    except:\n",
    "        return -9\n",
    "\n",
    "def combine_vcf_stats(filedir, prefix):\n",
    "    #hardy_files = !ls {filedir}/{prefix}*.hwe\n",
    "    print \"hardy files\"\n",
    "    files = os.listdir(filedir)\n",
    "    hardy_files = [os.path.join(filedir,f) for f in files if f.endswith('hwe')]\n",
    "    hardy = pd.read_csv(hardy_files[0], sep=\"\\t\")\n",
    "\n",
    "    hardy.columns = ['CHROM', 'POS', 'OBS(HOM1/HET/HOM2)', 'E(HOM1/HET/HOM2)', 'ChiSq_HWE',\n",
    "       'P_HWE', 'P_HET_DEFICIT', 'P_HET_EXCESS']\n",
    "    hardy.index = hardy.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    print \"loci files\"\n",
    "    #loci_files = !ls {filedir}/{prefix}*.l* | grep -v log\n",
    "    loci_files = [os.path.join(filedir,f) for f in files if '.l' in f and 'log' not in f]\n",
    "    loci_df = pd.concat([pd.read_csv(x, sep=\"\\t\", skiprows=0) for x in loci_files], axis=1)\n",
    "    chrom_pos = loci_df.ix[:,0:2]\n",
    "    \n",
    "    print \"frq files\"\n",
    "    #frq_files = !ls {filedir}/{prefix}*.frq* | grep -v count\n",
    "    frq_files = [os.path.join(filedir,f) for f in files if f.endswith('frq')]\n",
    "    frq_data = []\n",
    "    h = open(frq_files[0])\n",
    "    header = h.readline().strip().split()\n",
    "    for line in h:\n",
    "        frq_data.append(line.strip().split('\\t'))\n",
    "\n",
    "    header = ['CHROM', 'POS', 'N_ALLELES', 'N_CHR', 'A1_FREQ', \"A2_FREQ\"]\n",
    "    frq_df = pd.DataFrame(frq_data)\n",
    "    print(frq_df.columns)\n",
    "    #frq_df = frq_df.drop([6,7],axis=1)\n",
    "    frq_df.columns = header\n",
    "    frq_df.index = frq_df.apply(lambda x: \"%s-%s\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    print \"loci df\"\n",
    "    loci_df = loci_df.drop(['CHROM','CHR','POS'], axis=1)\n",
    "    loci_df = pd.concat([chrom_pos, loci_df], axis=1)\n",
    "    loci_df.index = loci_df.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    print \"identifying alleles\"\n",
    "    loci_df = pd.concat([loci_df, frq_df, hardy], axis=1)\n",
    "    loci_df[\"A1_allele\"] = loci_df.apply(lambda row: row.A1_FREQ.split(\":\")[0], axis=1)\n",
    "    loci_df[\"A2_allele\"] = loci_df.apply(lambda row: row.A2_FREQ.split(\":\")[0], axis=1)\n",
    "    \n",
    "    print \"calculating allele freqs\"\n",
    "    loci_df[\"A1_freq\"] = loci_df.apply(lambda row: float(row.A1_FREQ.split(\":\")[1]), axis=1)\n",
    "    loci_df[\"A2_freq\"] = loci_df.apply(lambda row: float(row.A2_FREQ.split(\":\")[1]), axis=1)\n",
    "    \n",
    "    print \"getting MAF\"\n",
    "    loci_df['MAF'] = loci_df.apply(get_MAF, axis=1)\n",
    "    loci_df = loci_df.drop(['CHROM', 'POS'], axis=1)\n",
    "    \n",
    "    print \"calculating FIS\"\n",
    "    loci_df['Fis'] = loci_df['OBS(HOM1/HET/HOM2)'].apply(calculate_Fis)\n",
    "    \n",
    "    print 'done'\n",
    "    return loci_df, frq_df, hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/lindb/wbp/concatenated/snps/'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hardy files\n",
      "loci files\n",
      "frq files\n",
      "Int64Index([0, 1, 2, 3, 4, 5], dtype='int64')\n",
      "finishing\n"
     ]
    }
   ],
   "source": [
    "loci_df, frq_df, hardy = combine_vcf_stats(snp_dir, \"samtools\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUM_DEPTH</th>\n",
       "      <th>SUMSQ_DEPTH</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "      <th>VAR_DEPTH</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPE_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>N_CHR</th>\n",
       "      <th>A1_FREQ</th>\n",
       "      <th>A2_FREQ</th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "      <th>E(HOM1/HET/HOM2)</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "      <th>A1_allele</th>\n",
       "      <th>A2_allele</th>\n",
       "      <th>A1_freq</th>\n",
       "      <th>A2_freq</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Fis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE_16_length_56_cov_1.375000-61</th>\n",
       "      <td>434</td>\n",
       "      <td>2138</td>\n",
       "      <td>1.77869</td>\n",
       "      <td>5.6216</td>\n",
       "      <td>591</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>212</td>\n",
       "      <td>0.434426</td>\n",
       "      <td>2</td>\n",
       "      <td>276</td>\n",
       "      <td>A:0</td>\n",
       "      <td>T:1</td>\n",
       "      <td>0/0/138</td>\n",
       "      <td>0.00/0.00/138.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_73_length_45_cov_1.955556-63</th>\n",
       "      <td>1063</td>\n",
       "      <td>9127</td>\n",
       "      <td>4.35656</td>\n",
       "      <td>18.5020</td>\n",
       "      <td>334</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>96</td>\n",
       "      <td>0.196721</td>\n",
       "      <td>2</td>\n",
       "      <td>392</td>\n",
       "      <td>T:0.168367</td>\n",
       "      <td>C:0.831633</td>\n",
       "      <td>17/32/147</td>\n",
       "      <td>5.56/54.89/135.56</td>\n",
       "      <td>34.080940</td>\n",
       "      <td>1.870726e-07</td>\n",
       "      <td>1.870726e-07</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.168367</td>\n",
       "      <td>0.831633</td>\n",
       "      <td>0.168367</td>\n",
       "      <td>0.418479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-22</th>\n",
       "      <td>17558</td>\n",
       "      <td>2525410</td>\n",
       "      <td>71.95900</td>\n",
       "      <td>5193.2200</td>\n",
       "      <td>999</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>0.081967</td>\n",
       "      <td>2</td>\n",
       "      <td>448</td>\n",
       "      <td>T:0.00892857</td>\n",
       "      <td>C:0.991071</td>\n",
       "      <td>0/4/220</td>\n",
       "      <td>0.02/3.96/220.02</td>\n",
       "      <td>0.018180</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.986592</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.008929</td>\n",
       "      <td>0.991071</td>\n",
       "      <td>0.008929</td>\n",
       "      <td>-0.006757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-45</th>\n",
       "      <td>2732</td>\n",
       "      <td>66896</td>\n",
       "      <td>11.19670</td>\n",
       "      <td>149.4100</td>\n",
       "      <td>999</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "      <td>0.176230</td>\n",
       "      <td>2</td>\n",
       "      <td>402</td>\n",
       "      <td>T:0.0348259</td>\n",
       "      <td>A:0.965174</td>\n",
       "      <td>0/14/187</td>\n",
       "      <td>0.24/13.51/187.24</td>\n",
       "      <td>0.261691</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.791139</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.034826</td>\n",
       "      <td>0.965174</td>\n",
       "      <td>0.034826</td>\n",
       "      <td>-0.033505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-46</th>\n",
       "      <td>2726</td>\n",
       "      <td>66620</td>\n",
       "      <td>11.17210</td>\n",
       "      <td>148.8260</td>\n",
       "      <td>999</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "      <td>0.176230</td>\n",
       "      <td>2</td>\n",
       "      <td>402</td>\n",
       "      <td>T:0.0174129</td>\n",
       "      <td>A:0.982587</td>\n",
       "      <td>0/7/194</td>\n",
       "      <td>0.06/6.88/194.06</td>\n",
       "      <td>0.063125</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.948286</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.017413</td>\n",
       "      <td>0.982587</td>\n",
       "      <td>0.017413</td>\n",
       "      <td>-0.015190</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    SUM_DEPTH  SUMSQ_DEPTH  MEAN_DEPTH  \\\n",
       "NODE_16_length_56_cov_1.375000-61         434         2138     1.77869   \n",
       "NODE_73_length_45_cov_1.955556-63        1063         9127     4.35656   \n",
       "NODE_117_length_99_cov_3.979798-22      17558      2525410    71.95900   \n",
       "NODE_117_length_99_cov_3.979798-45       2732        66896    11.19670   \n",
       "NODE_117_length_99_cov_3.979798-46       2726        66620    11.17210   \n",
       "\n",
       "                                    VAR_DEPTH  QUAL  N_DATA  \\\n",
       "NODE_16_length_56_cov_1.375000-61      5.6216   591     488   \n",
       "NODE_73_length_45_cov_1.955556-63     18.5020   334     488   \n",
       "NODE_117_length_99_cov_3.979798-22  5193.2200   999     488   \n",
       "NODE_117_length_99_cov_3.979798-45   149.4100   999     488   \n",
       "NODE_117_length_99_cov_3.979798-46   148.8260   999     488   \n",
       "\n",
       "                                    N_GENOTYPE_FILTERED  N_MISS    F_MISS  \\\n",
       "NODE_16_length_56_cov_1.375000-61                     0     212  0.434426   \n",
       "NODE_73_length_45_cov_1.955556-63                     0      96  0.196721   \n",
       "NODE_117_length_99_cov_3.979798-22                    0      40  0.081967   \n",
       "NODE_117_length_99_cov_3.979798-45                    0      86  0.176230   \n",
       "NODE_117_length_99_cov_3.979798-46                    0      86  0.176230   \n",
       "\n",
       "                                   N_ALLELES N_CHR       A1_FREQ     A2_FREQ  \\\n",
       "NODE_16_length_56_cov_1.375000-61          2   276           A:0         T:1   \n",
       "NODE_73_length_45_cov_1.955556-63          2   392    T:0.168367  C:0.831633   \n",
       "NODE_117_length_99_cov_3.979798-22         2   448  T:0.00892857  C:0.991071   \n",
       "NODE_117_length_99_cov_3.979798-45         2   402   T:0.0348259  A:0.965174   \n",
       "NODE_117_length_99_cov_3.979798-46         2   402   T:0.0174129  A:0.982587   \n",
       "\n",
       "                                   OBS(HOM1/HET/HOM2)   E(HOM1/HET/HOM2)  \\\n",
       "NODE_16_length_56_cov_1.375000-61             0/0/138   0.00/0.00/138.00   \n",
       "NODE_73_length_45_cov_1.955556-63           17/32/147  5.56/54.89/135.56   \n",
       "NODE_117_length_99_cov_3.979798-22            0/4/220   0.02/3.96/220.02   \n",
       "NODE_117_length_99_cov_3.979798-45           0/14/187  0.24/13.51/187.24   \n",
       "NODE_117_length_99_cov_3.979798-46            0/7/194   0.06/6.88/194.06   \n",
       "\n",
       "                                    ChiSq_HWE         P_HWE  P_HET_DEFICIT  \\\n",
       "NODE_16_length_56_cov_1.375000-61         NaN  1.000000e+00   1.000000e+00   \n",
       "NODE_73_length_45_cov_1.955556-63   34.080940  1.870726e-07   1.870726e-07   \n",
       "NODE_117_length_99_cov_3.979798-22   0.018180  1.000000e+00   1.000000e+00   \n",
       "NODE_117_length_99_cov_3.979798-45   0.261691  1.000000e+00   1.000000e+00   \n",
       "NODE_117_length_99_cov_3.979798-46   0.063125  1.000000e+00   1.000000e+00   \n",
       "\n",
       "                                    P_HET_EXCESS A1_allele A2_allele  \\\n",
       "NODE_16_length_56_cov_1.375000-61       1.000000         A         T   \n",
       "NODE_73_length_45_cov_1.955556-63       1.000000         T         C   \n",
       "NODE_117_length_99_cov_3.979798-22      0.986592         T         C   \n",
       "NODE_117_length_99_cov_3.979798-45      0.791139         T         A   \n",
       "NODE_117_length_99_cov_3.979798-46      0.948286         T         A   \n",
       "\n",
       "                                     A1_freq   A2_freq       MAF       Fis  \n",
       "NODE_16_length_56_cov_1.375000-61   0.000000  1.000000  0.000000       NaN  \n",
       "NODE_73_length_45_cov_1.955556-63   0.168367  0.831633  0.168367  0.418479  \n",
       "NODE_117_length_99_cov_3.979798-22  0.008929  0.991071  0.008929 -0.006757  \n",
       "NODE_117_length_99_cov_3.979798-45  0.034826  0.965174  0.034826 -0.033505  \n",
       "NODE_117_length_99_cov_3.979798-46  0.017413  0.982587  0.017413 -0.015190  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>N_CHR</th>\n",
       "      <th>A1_FREQ</th>\n",
       "      <th>A2_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE_16_length_56_cov_1.375000-61</th>\n",
       "      <td>NODE_16_length_56_cov_1.375000</td>\n",
       "      <td>61</td>\n",
       "      <td>2</td>\n",
       "      <td>276</td>\n",
       "      <td>A:0</td>\n",
       "      <td>T:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_73_length_45_cov_1.955556-63</th>\n",
       "      <td>NODE_73_length_45_cov_1.955556</td>\n",
       "      <td>63</td>\n",
       "      <td>2</td>\n",
       "      <td>392</td>\n",
       "      <td>T:0.168367</td>\n",
       "      <td>C:0.831633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-22</th>\n",
       "      <td>NODE_117_length_99_cov_3.979798</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>448</td>\n",
       "      <td>T:0.00892857</td>\n",
       "      <td>C:0.991071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-45</th>\n",
       "      <td>NODE_117_length_99_cov_3.979798</td>\n",
       "      <td>45</td>\n",
       "      <td>2</td>\n",
       "      <td>402</td>\n",
       "      <td>T:0.0348259</td>\n",
       "      <td>A:0.965174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-46</th>\n",
       "      <td>NODE_117_length_99_cov_3.979798</td>\n",
       "      <td>46</td>\n",
       "      <td>2</td>\n",
       "      <td>402</td>\n",
       "      <td>T:0.0174129</td>\n",
       "      <td>A:0.982587</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                              CHROM POS  \\\n",
       "NODE_16_length_56_cov_1.375000-61    NODE_16_length_56_cov_1.375000  61   \n",
       "NODE_73_length_45_cov_1.955556-63    NODE_73_length_45_cov_1.955556  63   \n",
       "NODE_117_length_99_cov_3.979798-22  NODE_117_length_99_cov_3.979798  22   \n",
       "NODE_117_length_99_cov_3.979798-45  NODE_117_length_99_cov_3.979798  45   \n",
       "NODE_117_length_99_cov_3.979798-46  NODE_117_length_99_cov_3.979798  46   \n",
       "\n",
       "                                   N_ALLELES N_CHR       A1_FREQ     A2_FREQ  \n",
       "NODE_16_length_56_cov_1.375000-61          2   276           A:0         T:1  \n",
       "NODE_73_length_45_cov_1.955556-63          2   392    T:0.168367  C:0.831633  \n",
       "NODE_117_length_99_cov_3.979798-22         2   448  T:0.00892857  C:0.991071  \n",
       "NODE_117_length_99_cov_3.979798-45         2   402   T:0.0348259  A:0.965174  \n",
       "NODE_117_length_99_cov_3.979798-46         2   402   T:0.0174129  A:0.982587  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frq_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "      <th>E(HOM1/HET/HOM2)</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE_16_length_56_cov_1.375000-61</th>\n",
       "      <td>NODE_16_length_56_cov_1.375000</td>\n",
       "      <td>61</td>\n",
       "      <td>0/0/138</td>\n",
       "      <td>0.00/0.00/138.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_73_length_45_cov_1.955556-63</th>\n",
       "      <td>NODE_73_length_45_cov_1.955556</td>\n",
       "      <td>63</td>\n",
       "      <td>17/32/147</td>\n",
       "      <td>5.56/54.89/135.56</td>\n",
       "      <td>34.080940</td>\n",
       "      <td>1.870726e-07</td>\n",
       "      <td>1.870726e-07</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-22</th>\n",
       "      <td>NODE_117_length_99_cov_3.979798</td>\n",
       "      <td>22</td>\n",
       "      <td>0/4/220</td>\n",
       "      <td>0.02/3.96/220.02</td>\n",
       "      <td>0.018180</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.986592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-45</th>\n",
       "      <td>NODE_117_length_99_cov_3.979798</td>\n",
       "      <td>45</td>\n",
       "      <td>0/14/187</td>\n",
       "      <td>0.24/13.51/187.24</td>\n",
       "      <td>0.261691</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.791139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-46</th>\n",
       "      <td>NODE_117_length_99_cov_3.979798</td>\n",
       "      <td>46</td>\n",
       "      <td>0/7/194</td>\n",
       "      <td>0.06/6.88/194.06</td>\n",
       "      <td>0.063125</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.948286</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                              CHROM  POS  \\\n",
       "NODE_16_length_56_cov_1.375000-61    NODE_16_length_56_cov_1.375000   61   \n",
       "NODE_73_length_45_cov_1.955556-63    NODE_73_length_45_cov_1.955556   63   \n",
       "NODE_117_length_99_cov_3.979798-22  NODE_117_length_99_cov_3.979798   22   \n",
       "NODE_117_length_99_cov_3.979798-45  NODE_117_length_99_cov_3.979798   45   \n",
       "NODE_117_length_99_cov_3.979798-46  NODE_117_length_99_cov_3.979798   46   \n",
       "\n",
       "                                   OBS(HOM1/HET/HOM2)   E(HOM1/HET/HOM2)  \\\n",
       "NODE_16_length_56_cov_1.375000-61             0/0/138   0.00/0.00/138.00   \n",
       "NODE_73_length_45_cov_1.955556-63           17/32/147  5.56/54.89/135.56   \n",
       "NODE_117_length_99_cov_3.979798-22            0/4/220   0.02/3.96/220.02   \n",
       "NODE_117_length_99_cov_3.979798-45           0/14/187  0.24/13.51/187.24   \n",
       "NODE_117_length_99_cov_3.979798-46            0/7/194   0.06/6.88/194.06   \n",
       "\n",
       "                                    ChiSq_HWE         P_HWE  P_HET_DEFICIT  \\\n",
       "NODE_16_length_56_cov_1.375000-61         NaN  1.000000e+00   1.000000e+00   \n",
       "NODE_73_length_45_cov_1.955556-63   34.080940  1.870726e-07   1.870726e-07   \n",
       "NODE_117_length_99_cov_3.979798-22   0.018180  1.000000e+00   1.000000e+00   \n",
       "NODE_117_length_99_cov_3.979798-45   0.261691  1.000000e+00   1.000000e+00   \n",
       "NODE_117_length_99_cov_3.979798-46   0.063125  1.000000e+00   1.000000e+00   \n",
       "\n",
       "                                    P_HET_EXCESS  \n",
       "NODE_16_length_56_cov_1.375000-61       1.000000  \n",
       "NODE_73_length_45_cov_1.955556-63       1.000000  \n",
       "NODE_117_length_99_cov_3.979798-22      0.986592  \n",
       "NODE_117_length_99_cov_3.979798-45      0.791139  \n",
       "NODE_117_length_99_cov_3.979798-46      0.948286  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hardy.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "loci_file = '/home/lindb/wbp/concatenated/snps/loci_vcf_stats.txt'\n",
    "loci_df.to_csv(loci_file,header=True,index=True,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "frq_file = '/home/lindb/wbp/concatenated/snps/frq_vcf_stats.txt'\n",
    "frq_df.to_csv(frq_file,header=True,index=True,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hardy_file = '/home/lindb/wbp/concatenated/snps/hwe_vcf_stats.txt'\n",
    "hardy.to_csv(hardy_file,header=True,index=True,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Impute with Beagle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    cd /home/lindb/wbp/concatenated/snps/beagle40\n",
    "    ln -s ../samtools_1.3.vcf.gz.recode.vcf.gz\n",
    "    /home/lindb/g/src/jdk1.8.0_73/bin/java -jar ~/g/src/BEAGLE4/beagle.r1399.jar \\\n",
    "    gl=samtools_1.3.vcf.gz.recode.vcf.gz \\\n",
    "    out=beagle40 \\\n",
    "    nthreads=32 \\\n",
    "    phase-its=20 \\\n",
    "    burnin-its=20 \\\n",
    "    impute-its=20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "beagle_dir = os.path.join(snp_dir, \"beagle40\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "beagle_vcf_gz = os.path.join(beagle_dir, \"beagle40.vcf.gz\")\n",
    "assert os.path.exists(beagle_vcf_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--depth\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 144.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--site-depth\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 158.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--site-mean-depth\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 155.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--site-quality\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Quality for Each Site\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 90.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--missing-indv\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Individual Missingness\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 173.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--missing-site\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Site Missingness\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 186.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--freq\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 178.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--counts\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 179.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--hardy\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting HWE statistics (but only for biallelic loci)\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 206.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--het\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Individual Heterozygosity\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 183.00 seconds\n"
     ]
    }
   ],
   "source": [
    "get_vcf_stats(beagle_vcf_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/lindb/wbp/concatenated/snps/beagle40'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beagle_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hardy files\n",
      "loci files\n",
      "frq files\n",
      "Int64Index([0, 1, 2, 3, 4, 5], dtype='int64')\n",
      "loci df\n",
      "identifying alleles\n",
      "calculating allele freqs\n",
      "getting MAF\n",
      "calculating FIS\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "loci_df_beagle, freq_df_beagle, hardy_beagle = combine_vcf_stats(beagle_dir, \"beagle40\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUM_DEPTH</th>\n",
       "      <th>SUMSQ_DEPTH</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "      <th>VAR_DEPTH</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPE_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>N_CHR</th>\n",
       "      <th>A1_FREQ</th>\n",
       "      <th>A2_FREQ</th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "      <th>E(HOM1/HET/HOM2)</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "      <th>A1_allele</th>\n",
       "      <th>A2_allele</th>\n",
       "      <th>A1_freq</th>\n",
       "      <th>A2_freq</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Fis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE_16_length_56_cov_1.375000-61</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0</td>\n",
       "      <td>T:1</td>\n",
       "      <td>0/0/244</td>\n",
       "      <td>0.00/0.00/244.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_73_length_45_cov_1.955556-63</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.0122951</td>\n",
       "      <td>C:0.987705</td>\n",
       "      <td>0/6/238</td>\n",
       "      <td>0.04/5.93/238.04</td>\n",
       "      <td>0.037809</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.693896e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.012295</td>\n",
       "      <td>0.987705</td>\n",
       "      <td>0.012295</td>\n",
       "      <td>-0.010373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-22</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.00204918</td>\n",
       "      <td>C:0.997951</td>\n",
       "      <td>0/1/243</td>\n",
       "      <td>0.00/1.00/243.00</td>\n",
       "      <td>0.001029</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.997951</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-45</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.0204918</td>\n",
       "      <td>A:0.979508</td>\n",
       "      <td>0/10/234</td>\n",
       "      <td>0.10/9.80/234.10</td>\n",
       "      <td>0.106791</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.102373e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.020492</td>\n",
       "      <td>0.979508</td>\n",
       "      <td>0.020492</td>\n",
       "      <td>-0.018828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-46</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.0204918</td>\n",
       "      <td>A:0.979508</td>\n",
       "      <td>0/10/234</td>\n",
       "      <td>0.10/9.80/234.10</td>\n",
       "      <td>0.106791</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.102373e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.020492</td>\n",
       "      <td>0.979508</td>\n",
       "      <td>0.020492</td>\n",
       "      <td>-0.018828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-64</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.0204918</td>\n",
       "      <td>G:0.979508</td>\n",
       "      <td>0/10/234</td>\n",
       "      <td>0.10/9.80/234.10</td>\n",
       "      <td>0.106791</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.102373e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.020492</td>\n",
       "      <td>0.979508</td>\n",
       "      <td>0.020492</td>\n",
       "      <td>-0.018828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-72</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.258197</td>\n",
       "      <td>C:0.741803</td>\n",
       "      <td>4/118/122</td>\n",
       "      <td>16.27/93.47/134.27</td>\n",
       "      <td>16.809900</td>\n",
       "      <td>1.834113e-05</td>\n",
       "      <td>9.999986e-01</td>\n",
       "      <td>1.048217e-05</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.258197</td>\n",
       "      <td>0.741803</td>\n",
       "      <td>0.258197</td>\n",
       "      <td>-0.259888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-82</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0.122951</td>\n",
       "      <td>G:0.877049</td>\n",
       "      <td>0/60/184</td>\n",
       "      <td>3.69/52.62/187.69</td>\n",
       "      <td>4.795179</td>\n",
       "      <td>3.172595e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.585835e-02</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.122951</td>\n",
       "      <td>0.877049</td>\n",
       "      <td>0.122951</td>\n",
       "      <td>-0.137850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-97</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.215164</td>\n",
       "      <td>C:0.784836</td>\n",
       "      <td>6/93/145</td>\n",
       "      <td>11.30/82.41/150.30</td>\n",
       "      <td>4.031129</td>\n",
       "      <td>5.683300e-02</td>\n",
       "      <td>9.887311e-01</td>\n",
       "      <td>3.198435e-02</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.215164</td>\n",
       "      <td>0.784836</td>\n",
       "      <td>0.215164</td>\n",
       "      <td>-0.126222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-98</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0.153689</td>\n",
       "      <td>C:0.846311</td>\n",
       "      <td>7/61/176</td>\n",
       "      <td>5.76/63.47/174.76</td>\n",
       "      <td>0.370495</td>\n",
       "      <td>4.699040e-01</td>\n",
       "      <td>3.337456e-01</td>\n",
       "      <td>8.160896e-01</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.153689</td>\n",
       "      <td>0.846311</td>\n",
       "      <td>0.153689</td>\n",
       "      <td>0.040936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-102</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.190574</td>\n",
       "      <td>C:0.809426</td>\n",
       "      <td>9/75/160</td>\n",
       "      <td>8.86/75.28/159.86</td>\n",
       "      <td>0.003295</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>5.363668e-01</td>\n",
       "      <td>6.263141e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.190574</td>\n",
       "      <td>0.809426</td>\n",
       "      <td>0.190574</td>\n",
       "      <td>0.005717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-106</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.471311</td>\n",
       "      <td>C:0.528689</td>\n",
       "      <td>32/166/46</td>\n",
       "      <td>54.20/121.60/68.20</td>\n",
       "      <td>32.533620</td>\n",
       "      <td>1.258374e-08</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.825594e-09</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.471311</td>\n",
       "      <td>0.528689</td>\n",
       "      <td>0.471311</td>\n",
       "      <td>-0.362353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-107</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0.502049</td>\n",
       "      <td>G:0.497951</td>\n",
       "      <td>35/175/34</td>\n",
       "      <td>61.50/122.00/60.50</td>\n",
       "      <td>46.054290</td>\n",
       "      <td>1.311706e-11</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>8.048139e-12</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.502049</td>\n",
       "      <td>0.497951</td>\n",
       "      <td>0.497951</td>\n",
       "      <td>-0.431511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_162_length_98_cov_5.010204-11</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.995902</td>\n",
       "      <td>C:0.00409836</td>\n",
       "      <td>242/2/0</td>\n",
       "      <td>242.00/1.99/0.00</td>\n",
       "      <td>0.004132</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.979466e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.995902</td>\n",
       "      <td>0.004098</td>\n",
       "      <td>0.004098</td>\n",
       "      <td>-0.002058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_162_length_98_cov_5.010204-13</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.997951</td>\n",
       "      <td>T:0.00204918</td>\n",
       "      <td>243/1/0</td>\n",
       "      <td>243.00/1.00/0.00</td>\n",
       "      <td>0.001029</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.997951</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_162_length_98_cov_5.010204-17</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>G:0.969262</td>\n",
       "      <td>A:0.0307377</td>\n",
       "      <td>229/15/0</td>\n",
       "      <td>229.23/14.54/0.23</td>\n",
       "      <td>0.245386</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>8.010857e-01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.969262</td>\n",
       "      <td>0.030738</td>\n",
       "      <td>0.030738</td>\n",
       "      <td>-0.029598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_162_length_98_cov_5.010204-24</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.997951</td>\n",
       "      <td>T:0.00204918</td>\n",
       "      <td>243/1/0</td>\n",
       "      <td>243.00/1.00/0.00</td>\n",
       "      <td>0.001029</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.997951</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_162_length_98_cov_5.010204-27</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.981557</td>\n",
       "      <td>T:0.0184426</td>\n",
       "      <td>235/9/0</td>\n",
       "      <td>235.08/8.83/0.08</td>\n",
       "      <td>0.086140</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.276674e-01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.981557</td>\n",
       "      <td>0.018443</td>\n",
       "      <td>0.018443</td>\n",
       "      <td>-0.016701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_162_length_98_cov_5.010204-28</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>G:0.983607</td>\n",
       "      <td>A:0.0163934</td>\n",
       "      <td>236/8/0</td>\n",
       "      <td>236.07/7.87/0.07</td>\n",
       "      <td>0.067778</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.433905e-01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.983607</td>\n",
       "      <td>0.016393</td>\n",
       "      <td>0.016393</td>\n",
       "      <td>-0.014583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_162_length_98_cov_5.010204-31</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.973361</td>\n",
       "      <td>A:0.0266393</td>\n",
       "      <td>231/13/0</td>\n",
       "      <td>231.17/12.65/0.17</td>\n",
       "      <td>0.182763</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>8.486940e-01</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.973361</td>\n",
       "      <td>0.026639</td>\n",
       "      <td>0.026639</td>\n",
       "      <td>-0.025263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_162_length_98_cov_5.010204-40</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>G:0.92418</td>\n",
       "      <td>A:0.0758197</td>\n",
       "      <td>207/37/0</td>\n",
       "      <td>208.40/34.19/1.40</td>\n",
       "      <td>1.642253</td>\n",
       "      <td>3.741141e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.283920e-01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.924180</td>\n",
       "      <td>0.075820</td>\n",
       "      <td>0.075820</td>\n",
       "      <td>-0.079823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_305_length_49_cov_1.204082-40</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0.0655738</td>\n",
       "      <td>T:0.934426</td>\n",
       "      <td>10/12/222</td>\n",
       "      <td>1.05/29.90/213.05</td>\n",
       "      <td>87.455160</td>\n",
       "      <td>2.083831e-10</td>\n",
       "      <td>2.083831e-10</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.065574</td>\n",
       "      <td>0.934426</td>\n",
       "      <td>0.065574</td>\n",
       "      <td>0.599507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_305_length_49_cov_1.204082-82</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0</td>\n",
       "      <td>C:1</td>\n",
       "      <td>0/0/244</td>\n",
       "      <td>0.00/0.00/244.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_411_length_91_cov_2.098901-11</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.911885</td>\n",
       "      <td>T:0.0881148</td>\n",
       "      <td>201/43/0</td>\n",
       "      <td>202.89/39.21/1.89</td>\n",
       "      <td>2.278278</td>\n",
       "      <td>2.305470e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.313331e-01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.911885</td>\n",
       "      <td>0.088115</td>\n",
       "      <td>0.088115</td>\n",
       "      <td>-0.094382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_411_length_91_cov_2.098901-18</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.944672</td>\n",
       "      <td>G:0.0553279</td>\n",
       "      <td>217/27/0</td>\n",
       "      <td>217.75/25.51/0.75</td>\n",
       "      <td>0.836981</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>4.672079e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.944672</td>\n",
       "      <td>0.055328</td>\n",
       "      <td>0.055328</td>\n",
       "      <td>-0.056399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_411_length_91_cov_2.098901-21</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.922131</td>\n",
       "      <td>C:0.0778689</td>\n",
       "      <td>206/38/0</td>\n",
       "      <td>207.48/35.04/1.48</td>\n",
       "      <td>1.739931</td>\n",
       "      <td>3.744188e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.096547e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.922131</td>\n",
       "      <td>0.077869</td>\n",
       "      <td>0.077869</td>\n",
       "      <td>-0.082222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_411_length_91_cov_2.098901-23</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>G:0.85041</td>\n",
       "      <td>T:0.14959</td>\n",
       "      <td>171/73/0</td>\n",
       "      <td>176.46/62.08/5.46</td>\n",
       "      <td>7.549868</td>\n",
       "      <td>3.858700e-03</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.733884e-03</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.850410</td>\n",
       "      <td>0.149590</td>\n",
       "      <td>0.149590</td>\n",
       "      <td>-0.173494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_411_length_91_cov_2.098901-25</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.868852</td>\n",
       "      <td>T:0.131148</td>\n",
       "      <td>184/56/4</td>\n",
       "      <td>184.20/55.61/4.20</td>\n",
       "      <td>0.012215</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>6.227237e-01</td>\n",
       "      <td>6.000197e-01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.868852</td>\n",
       "      <td>0.131148</td>\n",
       "      <td>0.131148</td>\n",
       "      <td>-0.005012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_411_length_91_cov_2.098901-32</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.868852</td>\n",
       "      <td>G:0.131148</td>\n",
       "      <td>184/56/4</td>\n",
       "      <td>184.20/55.61/4.20</td>\n",
       "      <td>0.012215</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>6.227237e-01</td>\n",
       "      <td>6.000197e-01</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.868852</td>\n",
       "      <td>0.131148</td>\n",
       "      <td>0.131148</td>\n",
       "      <td>-0.005012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_550_length_45_cov_3.133333-44</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.709016</td>\n",
       "      <td>T:0.290984</td>\n",
       "      <td>109/128/7</td>\n",
       "      <td>122.66/100.68/20.66</td>\n",
       "      <td>17.966010</td>\n",
       "      <td>1.094846e-05</td>\n",
       "      <td>9.999986e-01</td>\n",
       "      <td>8.184367e-06</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.709016</td>\n",
       "      <td>0.290984</td>\n",
       "      <td>0.290984</td>\n",
       "      <td>-0.268745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126184_length_92_cov_1.000000-35</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.915984</td>\n",
       "      <td>T:0.0840164</td>\n",
       "      <td>209/29/6</td>\n",
       "      <td>204.72/37.56/1.72</td>\n",
       "      <td>12.662520</td>\n",
       "      <td>3.047546e-03</td>\n",
       "      <td>3.047546e-03</td>\n",
       "      <td>9.995941e-01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.915984</td>\n",
       "      <td>0.084016</td>\n",
       "      <td>0.084016</td>\n",
       "      <td>0.229388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126184_length_92_cov_1.000000-42</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.94877</td>\n",
       "      <td>C:0.0512295</td>\n",
       "      <td>224/15/5</td>\n",
       "      <td>219.64/23.72/0.64</td>\n",
       "      <td>32.972130</td>\n",
       "      <td>9.742994e-05</td>\n",
       "      <td>9.742994e-05</td>\n",
       "      <td>9.999963e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.948770</td>\n",
       "      <td>0.051229</td>\n",
       "      <td>0.051229</td>\n",
       "      <td>0.368898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126184_length_92_cov_1.000000-44</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.764344</td>\n",
       "      <td>G:0.235656</td>\n",
       "      <td>146/81/17</td>\n",
       "      <td>142.55/87.90/13.55</td>\n",
       "      <td>1.503359</td>\n",
       "      <td>2.159177e-01</td>\n",
       "      <td>1.397860e-01</td>\n",
       "      <td>9.229479e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.764344</td>\n",
       "      <td>0.235656</td>\n",
       "      <td>0.235656</td>\n",
       "      <td>0.080382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126184_length_92_cov_1.000000-49</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0.827869</td>\n",
       "      <td>C:0.172131</td>\n",
       "      <td>174/56/14</td>\n",
       "      <td>167.23/69.54/7.23</td>\n",
       "      <td>9.251424</td>\n",
       "      <td>5.456015e-03</td>\n",
       "      <td>3.446771e-03</td>\n",
       "      <td>9.990364e-01</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.827869</td>\n",
       "      <td>0.172131</td>\n",
       "      <td>0.172131</td>\n",
       "      <td>0.196370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126184_length_92_cov_1.000000-50</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.94877</td>\n",
       "      <td>C:0.0512295</td>\n",
       "      <td>221/21/2</td>\n",
       "      <td>219.64/23.72/0.64</td>\n",
       "      <td>3.206931</td>\n",
       "      <td>1.198603e-01</td>\n",
       "      <td>1.198603e-01</td>\n",
       "      <td>9.822594e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.948770</td>\n",
       "      <td>0.051229</td>\n",
       "      <td>0.051229</td>\n",
       "      <td>0.116458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126184_length_92_cov_1.000000-53</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>G:0.971311</td>\n",
       "      <td>T:0.0286885</td>\n",
       "      <td>231/12/1</td>\n",
       "      <td>230.20/13.60/0.20</td>\n",
       "      <td>3.371058</td>\n",
       "      <td>1.745334e-01</td>\n",
       "      <td>1.745334e-01</td>\n",
       "      <td>9.880585e-01</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.971311</td>\n",
       "      <td>0.028688</td>\n",
       "      <td>0.028688</td>\n",
       "      <td>0.119349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126184_length_92_cov_1.000000-65</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0.944672</td>\n",
       "      <td>G:0.0553279</td>\n",
       "      <td>219/23/2</td>\n",
       "      <td>217.75/25.51/0.75</td>\n",
       "      <td>2.355664</td>\n",
       "      <td>1.566682e-01</td>\n",
       "      <td>1.566682e-01</td>\n",
       "      <td>9.721413e-01</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.944672</td>\n",
       "      <td>0.055328</td>\n",
       "      <td>0.055328</td>\n",
       "      <td>0.100104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126184_length_92_cov_1.000000-66</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>G:0.772541</td>\n",
       "      <td>A:0.227459</td>\n",
       "      <td>151/75/18</td>\n",
       "      <td>145.62/85.75/12.62</td>\n",
       "      <td>3.836041</td>\n",
       "      <td>6.638337e-02</td>\n",
       "      <td>3.747657e-02</td>\n",
       "      <td>9.832893e-01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.772541</td>\n",
       "      <td>0.227459</td>\n",
       "      <td>0.227459</td>\n",
       "      <td>0.127178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126184_length_92_cov_1.000000-76</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>G:0.981557</td>\n",
       "      <td>T:0.0184426</td>\n",
       "      <td>235/9/0</td>\n",
       "      <td>235.08/8.83/0.08</td>\n",
       "      <td>0.086140</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.276674e-01</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.981557</td>\n",
       "      <td>0.018443</td>\n",
       "      <td>0.018443</td>\n",
       "      <td>-0.016701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126184_length_92_cov_1.000000-77</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.895492</td>\n",
       "      <td>G:0.104508</td>\n",
       "      <td>202/33/9</td>\n",
       "      <td>195.66/45.67/2.66</td>\n",
       "      <td>18.779550</td>\n",
       "      <td>2.478658e-04</td>\n",
       "      <td>2.478658e-04</td>\n",
       "      <td>9.999687e-01</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.895492</td>\n",
       "      <td>0.104508</td>\n",
       "      <td>0.104508</td>\n",
       "      <td>0.278907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126184_length_92_cov_1.000000-80</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.989754</td>\n",
       "      <td>C:0.0102459</td>\n",
       "      <td>239/5/0</td>\n",
       "      <td>239.03/4.95/0.03</td>\n",
       "      <td>0.026148</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.795296e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.989754</td>\n",
       "      <td>0.010246</td>\n",
       "      <td>0.010246</td>\n",
       "      <td>-0.008282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126184_length_92_cov_1.000000-95</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>G:0.895492</td>\n",
       "      <td>A:0.104508</td>\n",
       "      <td>202/33/9</td>\n",
       "      <td>195.66/45.67/2.66</td>\n",
       "      <td>18.779550</td>\n",
       "      <td>2.478658e-04</td>\n",
       "      <td>2.478658e-04</td>\n",
       "      <td>9.999687e-01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.895492</td>\n",
       "      <td>0.104508</td>\n",
       "      <td>0.104508</td>\n",
       "      <td>0.278907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126204_length_90_cov_1.000000-45</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0</td>\n",
       "      <td>C:1</td>\n",
       "      <td>0/0/244</td>\n",
       "      <td>0.00/0.00/244.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126204_length_90_cov_1.000000-58</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0</td>\n",
       "      <td>G:1</td>\n",
       "      <td>0/0/244</td>\n",
       "      <td>0.00/0.00/244.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126204_length_90_cov_1.000000-67</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.0614754</td>\n",
       "      <td>C:0.938525</td>\n",
       "      <td>2/26/216</td>\n",
       "      <td>0.92/28.16/214.92</td>\n",
       "      <td>1.430368</td>\n",
       "      <td>2.215542e-01</td>\n",
       "      <td>2.215542e-01</td>\n",
       "      <td>9.497151e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.061475</td>\n",
       "      <td>0.938525</td>\n",
       "      <td>0.061475</td>\n",
       "      <td>0.078457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126204_length_90_cov_1.000000-76</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>G:0.329918</td>\n",
       "      <td>A:0.670082</td>\n",
       "      <td>21/119/104</td>\n",
       "      <td>26.56/107.88/109.56</td>\n",
       "      <td>2.590849</td>\n",
       "      <td>1.467330e-01</td>\n",
       "      <td>9.589912e-01</td>\n",
       "      <td>7.461743e-02</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.329918</td>\n",
       "      <td>0.670082</td>\n",
       "      <td>0.329918</td>\n",
       "      <td>-0.100784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126204_length_90_cov_1.000000-77</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0.172131</td>\n",
       "      <td>G:0.827869</td>\n",
       "      <td>12/60/172</td>\n",
       "      <td>7.23/69.54/167.23</td>\n",
       "      <td>4.592979</td>\n",
       "      <td>4.080076e-02</td>\n",
       "      <td>2.936487e-02</td>\n",
       "      <td>9.892343e-01</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.172131</td>\n",
       "      <td>0.827869</td>\n",
       "      <td>0.172131</td>\n",
       "      <td>0.138967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126213_length_59_cov_1.000000-14</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0</td>\n",
       "      <td>C:1</td>\n",
       "      <td>0/0/244</td>\n",
       "      <td>0.00/0.00/244.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126213_length_59_cov_1.000000-45</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0</td>\n",
       "      <td>C:1</td>\n",
       "      <td>0/0/244</td>\n",
       "      <td>0.00/0.00/244.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126213_length_59_cov_1.000000-59</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0</td>\n",
       "      <td>C:1</td>\n",
       "      <td>0/0/244</td>\n",
       "      <td>0.00/0.00/244.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126214_length_96_cov_1.000000-7</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.622951</td>\n",
       "      <td>T:0.377049</td>\n",
       "      <td>90/124/30</td>\n",
       "      <td>94.69/114.62/34.69</td>\n",
       "      <td>1.632973</td>\n",
       "      <td>2.226742e-01</td>\n",
       "      <td>9.172932e-01</td>\n",
       "      <td>1.331263e-01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.622951</td>\n",
       "      <td>0.377049</td>\n",
       "      <td>0.377049</td>\n",
       "      <td>-0.079591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126214_length_96_cov_1.000000-18</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.735656</td>\n",
       "      <td>G:0.264344</td>\n",
       "      <td>133/93/18</td>\n",
       "      <td>132.05/94.90/17.05</td>\n",
       "      <td>0.097764</td>\n",
       "      <td>7.431973e-01</td>\n",
       "      <td>4.237454e-01</td>\n",
       "      <td>6.981939e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.735656</td>\n",
       "      <td>0.264344</td>\n",
       "      <td>0.264344</td>\n",
       "      <td>0.022025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126214_length_96_cov_1.000000-63</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.067623</td>\n",
       "      <td>A:0.932377</td>\n",
       "      <td>4/25/215</td>\n",
       "      <td>1.12/30.77/212.12</td>\n",
       "      <td>8.576221</td>\n",
       "      <td>1.585400e-02</td>\n",
       "      <td>1.585400e-02</td>\n",
       "      <td>9.978919e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.067623</td>\n",
       "      <td>0.932377</td>\n",
       "      <td>0.067623</td>\n",
       "      <td>0.189144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126214_length_96_cov_1.000000-92</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.997951</td>\n",
       "      <td>C:0.00204918</td>\n",
       "      <td>243/1/0</td>\n",
       "      <td>243.00/1.00/0.00</td>\n",
       "      <td>0.001029</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.997951</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126214_length_96_cov_1.000000-99</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0.336066</td>\n",
       "      <td>G:0.663934</td>\n",
       "      <td>38/88/118</td>\n",
       "      <td>27.56/108.89/107.56</td>\n",
       "      <td>8.976994</td>\n",
       "      <td>3.872338e-03</td>\n",
       "      <td>2.083067e-03</td>\n",
       "      <td>9.991796e-01</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.336066</td>\n",
       "      <td>0.663934</td>\n",
       "      <td>0.336066</td>\n",
       "      <td>0.193466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126214_length_96_cov_1.000000-103</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>G:0.370902</td>\n",
       "      <td>A:0.629098</td>\n",
       "      <td>42/97/105</td>\n",
       "      <td>33.57/113.87/96.57</td>\n",
       "      <td>5.353787</td>\n",
       "      <td>2.028657e-02</td>\n",
       "      <td>1.375216e-02</td>\n",
       "      <td>9.933807e-01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.370902</td>\n",
       "      <td>0.629098</td>\n",
       "      <td>0.370902</td>\n",
       "      <td>0.149873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126217_length_62_cov_1.000000-28</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:1</td>\n",
       "      <td>T:0</td>\n",
       "      <td>244/0/0</td>\n",
       "      <td>244.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126222_length_45_cov_2.000000-70</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0</td>\n",
       "      <td>G:1</td>\n",
       "      <td>0/0/244</td>\n",
       "      <td>0.00/0.00/244.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126233_length_45_cov_1.000000-20</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0</td>\n",
       "      <td>G:1</td>\n",
       "      <td>0/0/244</td>\n",
       "      <td>0.00/0.00/244.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_2126233_length_45_cov_1.000000-45</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>G:0</td>\n",
       "      <td>C:1</td>\n",
       "      <td>0/0/244</td>\n",
       "      <td>0.00/0.00/244.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1300961 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         SUM_DEPTH  SUMSQ_DEPTH  MEAN_DEPTH  \\\n",
       "NODE_16_length_56_cov_1.375000-61                0            0         NaN   \n",
       "NODE_73_length_45_cov_1.955556-63                0            0         NaN   \n",
       "NODE_117_length_99_cov_3.979798-22               0            0         NaN   \n",
       "NODE_117_length_99_cov_3.979798-45               0            0         NaN   \n",
       "NODE_117_length_99_cov_3.979798-46               0            0         NaN   \n",
       "NODE_117_length_99_cov_3.979798-64               0            0         NaN   \n",
       "NODE_117_length_99_cov_3.979798-72               0            0         NaN   \n",
       "NODE_117_length_99_cov_3.979798-82               0            0         NaN   \n",
       "NODE_117_length_99_cov_3.979798-97               0            0         NaN   \n",
       "NODE_117_length_99_cov_3.979798-98               0            0         NaN   \n",
       "NODE_117_length_99_cov_3.979798-102              0            0         NaN   \n",
       "NODE_117_length_99_cov_3.979798-106              0            0         NaN   \n",
       "NODE_117_length_99_cov_3.979798-107              0            0         NaN   \n",
       "NODE_162_length_98_cov_5.010204-11               0            0         NaN   \n",
       "NODE_162_length_98_cov_5.010204-13               0            0         NaN   \n",
       "NODE_162_length_98_cov_5.010204-17               0            0         NaN   \n",
       "NODE_162_length_98_cov_5.010204-24               0            0         NaN   \n",
       "NODE_162_length_98_cov_5.010204-27               0            0         NaN   \n",
       "NODE_162_length_98_cov_5.010204-28               0            0         NaN   \n",
       "NODE_162_length_98_cov_5.010204-31               0            0         NaN   \n",
       "NODE_162_length_98_cov_5.010204-40               0            0         NaN   \n",
       "NODE_305_length_49_cov_1.204082-40               0            0         NaN   \n",
       "NODE_305_length_49_cov_1.204082-82               0            0         NaN   \n",
       "NODE_411_length_91_cov_2.098901-11               0            0         NaN   \n",
       "NODE_411_length_91_cov_2.098901-18               0            0         NaN   \n",
       "NODE_411_length_91_cov_2.098901-21               0            0         NaN   \n",
       "NODE_411_length_91_cov_2.098901-23               0            0         NaN   \n",
       "NODE_411_length_91_cov_2.098901-25               0            0         NaN   \n",
       "NODE_411_length_91_cov_2.098901-32               0            0         NaN   \n",
       "NODE_550_length_45_cov_3.133333-44               0            0         NaN   \n",
       "...                                            ...          ...         ...   \n",
       "NODE_2126184_length_92_cov_1.000000-35           0            0         NaN   \n",
       "NODE_2126184_length_92_cov_1.000000-42           0            0         NaN   \n",
       "NODE_2126184_length_92_cov_1.000000-44           0            0         NaN   \n",
       "NODE_2126184_length_92_cov_1.000000-49           0            0         NaN   \n",
       "NODE_2126184_length_92_cov_1.000000-50           0            0         NaN   \n",
       "NODE_2126184_length_92_cov_1.000000-53           0            0         NaN   \n",
       "NODE_2126184_length_92_cov_1.000000-65           0            0         NaN   \n",
       "NODE_2126184_length_92_cov_1.000000-66           0            0         NaN   \n",
       "NODE_2126184_length_92_cov_1.000000-76           0            0         NaN   \n",
       "NODE_2126184_length_92_cov_1.000000-77           0            0         NaN   \n",
       "NODE_2126184_length_92_cov_1.000000-80           0            0         NaN   \n",
       "NODE_2126184_length_92_cov_1.000000-95           0            0         NaN   \n",
       "NODE_2126204_length_90_cov_1.000000-45           0            0         NaN   \n",
       "NODE_2126204_length_90_cov_1.000000-58           0            0         NaN   \n",
       "NODE_2126204_length_90_cov_1.000000-67           0            0         NaN   \n",
       "NODE_2126204_length_90_cov_1.000000-76           0            0         NaN   \n",
       "NODE_2126204_length_90_cov_1.000000-77           0            0         NaN   \n",
       "NODE_2126213_length_59_cov_1.000000-14           0            0         NaN   \n",
       "NODE_2126213_length_59_cov_1.000000-45           0            0         NaN   \n",
       "NODE_2126213_length_59_cov_1.000000-59           0            0         NaN   \n",
       "NODE_2126214_length_96_cov_1.000000-7            0            0         NaN   \n",
       "NODE_2126214_length_96_cov_1.000000-18           0            0         NaN   \n",
       "NODE_2126214_length_96_cov_1.000000-63           0            0         NaN   \n",
       "NODE_2126214_length_96_cov_1.000000-92           0            0         NaN   \n",
       "NODE_2126214_length_96_cov_1.000000-99           0            0         NaN   \n",
       "NODE_2126214_length_96_cov_1.000000-103          0            0         NaN   \n",
       "NODE_2126217_length_62_cov_1.000000-28           0            0         NaN   \n",
       "NODE_2126222_length_45_cov_2.000000-70           0            0         NaN   \n",
       "NODE_2126233_length_45_cov_1.000000-20           0            0         NaN   \n",
       "NODE_2126233_length_45_cov_1.000000-45           0            0         NaN   \n",
       "\n",
       "                                         VAR_DEPTH  QUAL  N_DATA  \\\n",
       "NODE_16_length_56_cov_1.375000-61              NaN    -1     488   \n",
       "NODE_73_length_45_cov_1.955556-63              NaN    -1     488   \n",
       "NODE_117_length_99_cov_3.979798-22             NaN    -1     488   \n",
       "NODE_117_length_99_cov_3.979798-45             NaN    -1     488   \n",
       "NODE_117_length_99_cov_3.979798-46             NaN    -1     488   \n",
       "NODE_117_length_99_cov_3.979798-64             NaN    -1     488   \n",
       "NODE_117_length_99_cov_3.979798-72             NaN    -1     488   \n",
       "NODE_117_length_99_cov_3.979798-82             NaN    -1     488   \n",
       "NODE_117_length_99_cov_3.979798-97             NaN    -1     488   \n",
       "NODE_117_length_99_cov_3.979798-98             NaN    -1     488   \n",
       "NODE_117_length_99_cov_3.979798-102            NaN    -1     488   \n",
       "NODE_117_length_99_cov_3.979798-106            NaN    -1     488   \n",
       "NODE_117_length_99_cov_3.979798-107            NaN    -1     488   \n",
       "NODE_162_length_98_cov_5.010204-11             NaN    -1     488   \n",
       "NODE_162_length_98_cov_5.010204-13             NaN    -1     488   \n",
       "NODE_162_length_98_cov_5.010204-17             NaN    -1     488   \n",
       "NODE_162_length_98_cov_5.010204-24             NaN    -1     488   \n",
       "NODE_162_length_98_cov_5.010204-27             NaN    -1     488   \n",
       "NODE_162_length_98_cov_5.010204-28             NaN    -1     488   \n",
       "NODE_162_length_98_cov_5.010204-31             NaN    -1     488   \n",
       "NODE_162_length_98_cov_5.010204-40             NaN    -1     488   \n",
       "NODE_305_length_49_cov_1.204082-40             NaN    -1     488   \n",
       "NODE_305_length_49_cov_1.204082-82             NaN    -1     488   \n",
       "NODE_411_length_91_cov_2.098901-11             NaN    -1     488   \n",
       "NODE_411_length_91_cov_2.098901-18             NaN    -1     488   \n",
       "NODE_411_length_91_cov_2.098901-21             NaN    -1     488   \n",
       "NODE_411_length_91_cov_2.098901-23             NaN    -1     488   \n",
       "NODE_411_length_91_cov_2.098901-25             NaN    -1     488   \n",
       "NODE_411_length_91_cov_2.098901-32             NaN    -1     488   \n",
       "NODE_550_length_45_cov_3.133333-44             NaN    -1     488   \n",
       "...                                            ...   ...     ...   \n",
       "NODE_2126184_length_92_cov_1.000000-35         NaN    -1     488   \n",
       "NODE_2126184_length_92_cov_1.000000-42         NaN    -1     488   \n",
       "NODE_2126184_length_92_cov_1.000000-44         NaN    -1     488   \n",
       "NODE_2126184_length_92_cov_1.000000-49         NaN    -1     488   \n",
       "NODE_2126184_length_92_cov_1.000000-50         NaN    -1     488   \n",
       "NODE_2126184_length_92_cov_1.000000-53         NaN    -1     488   \n",
       "NODE_2126184_length_92_cov_1.000000-65         NaN    -1     488   \n",
       "NODE_2126184_length_92_cov_1.000000-66         NaN    -1     488   \n",
       "NODE_2126184_length_92_cov_1.000000-76         NaN    -1     488   \n",
       "NODE_2126184_length_92_cov_1.000000-77         NaN    -1     488   \n",
       "NODE_2126184_length_92_cov_1.000000-80         NaN    -1     488   \n",
       "NODE_2126184_length_92_cov_1.000000-95         NaN    -1     488   \n",
       "NODE_2126204_length_90_cov_1.000000-45         NaN    -1     488   \n",
       "NODE_2126204_length_90_cov_1.000000-58         NaN    -1     488   \n",
       "NODE_2126204_length_90_cov_1.000000-67         NaN    -1     488   \n",
       "NODE_2126204_length_90_cov_1.000000-76         NaN    -1     488   \n",
       "NODE_2126204_length_90_cov_1.000000-77         NaN    -1     488   \n",
       "NODE_2126213_length_59_cov_1.000000-14         NaN    -1     488   \n",
       "NODE_2126213_length_59_cov_1.000000-45         NaN    -1     488   \n",
       "NODE_2126213_length_59_cov_1.000000-59         NaN    -1     488   \n",
       "NODE_2126214_length_96_cov_1.000000-7          NaN    -1     488   \n",
       "NODE_2126214_length_96_cov_1.000000-18         NaN    -1     488   \n",
       "NODE_2126214_length_96_cov_1.000000-63         NaN    -1     488   \n",
       "NODE_2126214_length_96_cov_1.000000-92         NaN    -1     488   \n",
       "NODE_2126214_length_96_cov_1.000000-99         NaN    -1     488   \n",
       "NODE_2126214_length_96_cov_1.000000-103        NaN    -1     488   \n",
       "NODE_2126217_length_62_cov_1.000000-28         NaN    -1     488   \n",
       "NODE_2126222_length_45_cov_2.000000-70         NaN    -1     488   \n",
       "NODE_2126233_length_45_cov_1.000000-20         NaN    -1     488   \n",
       "NODE_2126233_length_45_cov_1.000000-45         NaN    -1     488   \n",
       "\n",
       "                                         N_GENOTYPE_FILTERED  N_MISS  F_MISS  \\\n",
       "NODE_16_length_56_cov_1.375000-61                          0       0       0   \n",
       "NODE_73_length_45_cov_1.955556-63                          0       0       0   \n",
       "NODE_117_length_99_cov_3.979798-22                         0       0       0   \n",
       "NODE_117_length_99_cov_3.979798-45                         0       0       0   \n",
       "NODE_117_length_99_cov_3.979798-46                         0       0       0   \n",
       "NODE_117_length_99_cov_3.979798-64                         0       0       0   \n",
       "NODE_117_length_99_cov_3.979798-72                         0       0       0   \n",
       "NODE_117_length_99_cov_3.979798-82                         0       0       0   \n",
       "NODE_117_length_99_cov_3.979798-97                         0       0       0   \n",
       "NODE_117_length_99_cov_3.979798-98                         0       0       0   \n",
       "NODE_117_length_99_cov_3.979798-102                        0       0       0   \n",
       "NODE_117_length_99_cov_3.979798-106                        0       0       0   \n",
       "NODE_117_length_99_cov_3.979798-107                        0       0       0   \n",
       "NODE_162_length_98_cov_5.010204-11                         0       0       0   \n",
       "NODE_162_length_98_cov_5.010204-13                         0       0       0   \n",
       "NODE_162_length_98_cov_5.010204-17                         0       0       0   \n",
       "NODE_162_length_98_cov_5.010204-24                         0       0       0   \n",
       "NODE_162_length_98_cov_5.010204-27                         0       0       0   \n",
       "NODE_162_length_98_cov_5.010204-28                         0       0       0   \n",
       "NODE_162_length_98_cov_5.010204-31                         0       0       0   \n",
       "NODE_162_length_98_cov_5.010204-40                         0       0       0   \n",
       "NODE_305_length_49_cov_1.204082-40                         0       0       0   \n",
       "NODE_305_length_49_cov_1.204082-82                         0       0       0   \n",
       "NODE_411_length_91_cov_2.098901-11                         0       0       0   \n",
       "NODE_411_length_91_cov_2.098901-18                         0       0       0   \n",
       "NODE_411_length_91_cov_2.098901-21                         0       0       0   \n",
       "NODE_411_length_91_cov_2.098901-23                         0       0       0   \n",
       "NODE_411_length_91_cov_2.098901-25                         0       0       0   \n",
       "NODE_411_length_91_cov_2.098901-32                         0       0       0   \n",
       "NODE_550_length_45_cov_3.133333-44                         0       0       0   \n",
       "...                                                      ...     ...     ...   \n",
       "NODE_2126184_length_92_cov_1.000000-35                     0       0       0   \n",
       "NODE_2126184_length_92_cov_1.000000-42                     0       0       0   \n",
       "NODE_2126184_length_92_cov_1.000000-44                     0       0       0   \n",
       "NODE_2126184_length_92_cov_1.000000-49                     0       0       0   \n",
       "NODE_2126184_length_92_cov_1.000000-50                     0       0       0   \n",
       "NODE_2126184_length_92_cov_1.000000-53                     0       0       0   \n",
       "NODE_2126184_length_92_cov_1.000000-65                     0       0       0   \n",
       "NODE_2126184_length_92_cov_1.000000-66                     0       0       0   \n",
       "NODE_2126184_length_92_cov_1.000000-76                     0       0       0   \n",
       "NODE_2126184_length_92_cov_1.000000-77                     0       0       0   \n",
       "NODE_2126184_length_92_cov_1.000000-80                     0       0       0   \n",
       "NODE_2126184_length_92_cov_1.000000-95                     0       0       0   \n",
       "NODE_2126204_length_90_cov_1.000000-45                     0       0       0   \n",
       "NODE_2126204_length_90_cov_1.000000-58                     0       0       0   \n",
       "NODE_2126204_length_90_cov_1.000000-67                     0       0       0   \n",
       "NODE_2126204_length_90_cov_1.000000-76                     0       0       0   \n",
       "NODE_2126204_length_90_cov_1.000000-77                     0       0       0   \n",
       "NODE_2126213_length_59_cov_1.000000-14                     0       0       0   \n",
       "NODE_2126213_length_59_cov_1.000000-45                     0       0       0   \n",
       "NODE_2126213_length_59_cov_1.000000-59                     0       0       0   \n",
       "NODE_2126214_length_96_cov_1.000000-7                      0       0       0   \n",
       "NODE_2126214_length_96_cov_1.000000-18                     0       0       0   \n",
       "NODE_2126214_length_96_cov_1.000000-63                     0       0       0   \n",
       "NODE_2126214_length_96_cov_1.000000-92                     0       0       0   \n",
       "NODE_2126214_length_96_cov_1.000000-99                     0       0       0   \n",
       "NODE_2126214_length_96_cov_1.000000-103                    0       0       0   \n",
       "NODE_2126217_length_62_cov_1.000000-28                     0       0       0   \n",
       "NODE_2126222_length_45_cov_2.000000-70                     0       0       0   \n",
       "NODE_2126233_length_45_cov_1.000000-20                     0       0       0   \n",
       "NODE_2126233_length_45_cov_1.000000-45                     0       0       0   \n",
       "\n",
       "                                        N_ALLELES N_CHR       A1_FREQ  \\\n",
       "NODE_16_length_56_cov_1.375000-61               2   488           A:0   \n",
       "NODE_73_length_45_cov_1.955556-63               2   488   T:0.0122951   \n",
       "NODE_117_length_99_cov_3.979798-22              2   488  T:0.00204918   \n",
       "NODE_117_length_99_cov_3.979798-45              2   488   T:0.0204918   \n",
       "NODE_117_length_99_cov_3.979798-46              2   488   T:0.0204918   \n",
       "NODE_117_length_99_cov_3.979798-64              2   488   T:0.0204918   \n",
       "NODE_117_length_99_cov_3.979798-72              2   488    T:0.258197   \n",
       "NODE_117_length_99_cov_3.979798-82              2   488    A:0.122951   \n",
       "NODE_117_length_99_cov_3.979798-97              2   488    T:0.215164   \n",
       "NODE_117_length_99_cov_3.979798-98              2   488    A:0.153689   \n",
       "NODE_117_length_99_cov_3.979798-102             2   488    T:0.190574   \n",
       "NODE_117_length_99_cov_3.979798-106             2   488    T:0.471311   \n",
       "NODE_117_length_99_cov_3.979798-107             2   488    A:0.502049   \n",
       "NODE_162_length_98_cov_5.010204-11              2   488    T:0.995902   \n",
       "NODE_162_length_98_cov_5.010204-13              2   488    C:0.997951   \n",
       "NODE_162_length_98_cov_5.010204-17              2   488    G:0.969262   \n",
       "NODE_162_length_98_cov_5.010204-24              2   488    C:0.997951   \n",
       "NODE_162_length_98_cov_5.010204-27              2   488    C:0.981557   \n",
       "NODE_162_length_98_cov_5.010204-28              2   488    G:0.983607   \n",
       "NODE_162_length_98_cov_5.010204-31              2   488    C:0.973361   \n",
       "NODE_162_length_98_cov_5.010204-40              2   488     G:0.92418   \n",
       "NODE_305_length_49_cov_1.204082-40              2   488   A:0.0655738   \n",
       "NODE_305_length_49_cov_1.204082-82              2   488           A:0   \n",
       "NODE_411_length_91_cov_2.098901-11              2   488    C:0.911885   \n",
       "NODE_411_length_91_cov_2.098901-18              2   488    T:0.944672   \n",
       "NODE_411_length_91_cov_2.098901-21              2   488    T:0.922131   \n",
       "NODE_411_length_91_cov_2.098901-23              2   488     G:0.85041   \n",
       "NODE_411_length_91_cov_2.098901-25              2   488    C:0.868852   \n",
       "NODE_411_length_91_cov_2.098901-32              2   488    C:0.868852   \n",
       "NODE_550_length_45_cov_3.133333-44              2   488    C:0.709016   \n",
       "...                                           ...   ...           ...   \n",
       "NODE_2126184_length_92_cov_1.000000-35          2   488    C:0.915984   \n",
       "NODE_2126184_length_92_cov_1.000000-42          2   488     T:0.94877   \n",
       "NODE_2126184_length_92_cov_1.000000-44          2   488    T:0.764344   \n",
       "NODE_2126184_length_92_cov_1.000000-49          2   488    A:0.827869   \n",
       "NODE_2126184_length_92_cov_1.000000-50          2   488     T:0.94877   \n",
       "NODE_2126184_length_92_cov_1.000000-53          2   488    G:0.971311   \n",
       "NODE_2126184_length_92_cov_1.000000-65          2   488    A:0.944672   \n",
       "NODE_2126184_length_92_cov_1.000000-66          2   488    G:0.772541   \n",
       "NODE_2126184_length_92_cov_1.000000-76          2   488    G:0.981557   \n",
       "NODE_2126184_length_92_cov_1.000000-77          2   488    C:0.895492   \n",
       "NODE_2126184_length_92_cov_1.000000-80          2   488    T:0.989754   \n",
       "NODE_2126184_length_92_cov_1.000000-95          2   488    G:0.895492   \n",
       "NODE_2126204_length_90_cov_1.000000-45          2   488           T:0   \n",
       "NODE_2126204_length_90_cov_1.000000-58          2   488           C:0   \n",
       "NODE_2126204_length_90_cov_1.000000-67          2   488   T:0.0614754   \n",
       "NODE_2126204_length_90_cov_1.000000-76          2   488    G:0.329918   \n",
       "NODE_2126204_length_90_cov_1.000000-77          2   488    A:0.172131   \n",
       "NODE_2126213_length_59_cov_1.000000-14          2   488           A:0   \n",
       "NODE_2126213_length_59_cov_1.000000-45          2   488           T:0   \n",
       "NODE_2126213_length_59_cov_1.000000-59          2   488           T:0   \n",
       "NODE_2126214_length_96_cov_1.000000-7           2   488    C:0.622951   \n",
       "NODE_2126214_length_96_cov_1.000000-18          2   488    T:0.735656   \n",
       "NODE_2126214_length_96_cov_1.000000-63          2   488    T:0.067623   \n",
       "NODE_2126214_length_96_cov_1.000000-92          2   488    T:0.997951   \n",
       "NODE_2126214_length_96_cov_1.000000-99          2   488    A:0.336066   \n",
       "NODE_2126214_length_96_cov_1.000000-103         2   488    G:0.370902   \n",
       "NODE_2126217_length_62_cov_1.000000-28          2   488           A:1   \n",
       "NODE_2126222_length_45_cov_2.000000-70          2   488           A:0   \n",
       "NODE_2126233_length_45_cov_1.000000-20          2   488           A:0   \n",
       "NODE_2126233_length_45_cov_1.000000-45          2   488           G:0   \n",
       "\n",
       "                                              A2_FREQ OBS(HOM1/HET/HOM2)  \\\n",
       "NODE_16_length_56_cov_1.375000-61                 T:1            0/0/244   \n",
       "NODE_73_length_45_cov_1.955556-63          C:0.987705            0/6/238   \n",
       "NODE_117_length_99_cov_3.979798-22         C:0.997951            0/1/243   \n",
       "NODE_117_length_99_cov_3.979798-45         A:0.979508           0/10/234   \n",
       "NODE_117_length_99_cov_3.979798-46         A:0.979508           0/10/234   \n",
       "NODE_117_length_99_cov_3.979798-64         G:0.979508           0/10/234   \n",
       "NODE_117_length_99_cov_3.979798-72         C:0.741803          4/118/122   \n",
       "NODE_117_length_99_cov_3.979798-82         G:0.877049           0/60/184   \n",
       "NODE_117_length_99_cov_3.979798-97         C:0.784836           6/93/145   \n",
       "NODE_117_length_99_cov_3.979798-98         C:0.846311           7/61/176   \n",
       "NODE_117_length_99_cov_3.979798-102        C:0.809426           9/75/160   \n",
       "NODE_117_length_99_cov_3.979798-106        C:0.528689          32/166/46   \n",
       "NODE_117_length_99_cov_3.979798-107        G:0.497951          35/175/34   \n",
       "NODE_162_length_98_cov_5.010204-11       C:0.00409836            242/2/0   \n",
       "NODE_162_length_98_cov_5.010204-13       T:0.00204918            243/1/0   \n",
       "NODE_162_length_98_cov_5.010204-17        A:0.0307377           229/15/0   \n",
       "NODE_162_length_98_cov_5.010204-24       T:0.00204918            243/1/0   \n",
       "NODE_162_length_98_cov_5.010204-27        T:0.0184426            235/9/0   \n",
       "NODE_162_length_98_cov_5.010204-28        A:0.0163934            236/8/0   \n",
       "NODE_162_length_98_cov_5.010204-31        A:0.0266393           231/13/0   \n",
       "NODE_162_length_98_cov_5.010204-40        A:0.0758197           207/37/0   \n",
       "NODE_305_length_49_cov_1.204082-40         T:0.934426          10/12/222   \n",
       "NODE_305_length_49_cov_1.204082-82                C:1            0/0/244   \n",
       "NODE_411_length_91_cov_2.098901-11        T:0.0881148           201/43/0   \n",
       "NODE_411_length_91_cov_2.098901-18        G:0.0553279           217/27/0   \n",
       "NODE_411_length_91_cov_2.098901-21        C:0.0778689           206/38/0   \n",
       "NODE_411_length_91_cov_2.098901-23          T:0.14959           171/73/0   \n",
       "NODE_411_length_91_cov_2.098901-25         T:0.131148           184/56/4   \n",
       "NODE_411_length_91_cov_2.098901-32         G:0.131148           184/56/4   \n",
       "NODE_550_length_45_cov_3.133333-44         T:0.290984          109/128/7   \n",
       "...                                               ...                ...   \n",
       "NODE_2126184_length_92_cov_1.000000-35    T:0.0840164           209/29/6   \n",
       "NODE_2126184_length_92_cov_1.000000-42    C:0.0512295           224/15/5   \n",
       "NODE_2126184_length_92_cov_1.000000-44     G:0.235656          146/81/17   \n",
       "NODE_2126184_length_92_cov_1.000000-49     C:0.172131          174/56/14   \n",
       "NODE_2126184_length_92_cov_1.000000-50    C:0.0512295           221/21/2   \n",
       "NODE_2126184_length_92_cov_1.000000-53    T:0.0286885           231/12/1   \n",
       "NODE_2126184_length_92_cov_1.000000-65    G:0.0553279           219/23/2   \n",
       "NODE_2126184_length_92_cov_1.000000-66     A:0.227459          151/75/18   \n",
       "NODE_2126184_length_92_cov_1.000000-76    T:0.0184426            235/9/0   \n",
       "NODE_2126184_length_92_cov_1.000000-77     G:0.104508           202/33/9   \n",
       "NODE_2126184_length_92_cov_1.000000-80    C:0.0102459            239/5/0   \n",
       "NODE_2126184_length_92_cov_1.000000-95     A:0.104508           202/33/9   \n",
       "NODE_2126204_length_90_cov_1.000000-45            C:1            0/0/244   \n",
       "NODE_2126204_length_90_cov_1.000000-58            G:1            0/0/244   \n",
       "NODE_2126204_length_90_cov_1.000000-67     C:0.938525           2/26/216   \n",
       "NODE_2126204_length_90_cov_1.000000-76     A:0.670082         21/119/104   \n",
       "NODE_2126204_length_90_cov_1.000000-77     G:0.827869          12/60/172   \n",
       "NODE_2126213_length_59_cov_1.000000-14            C:1            0/0/244   \n",
       "NODE_2126213_length_59_cov_1.000000-45            C:1            0/0/244   \n",
       "NODE_2126213_length_59_cov_1.000000-59            C:1            0/0/244   \n",
       "NODE_2126214_length_96_cov_1.000000-7      T:0.377049          90/124/30   \n",
       "NODE_2126214_length_96_cov_1.000000-18     G:0.264344          133/93/18   \n",
       "NODE_2126214_length_96_cov_1.000000-63     A:0.932377           4/25/215   \n",
       "NODE_2126214_length_96_cov_1.000000-92   C:0.00204918            243/1/0   \n",
       "NODE_2126214_length_96_cov_1.000000-99     G:0.663934          38/88/118   \n",
       "NODE_2126214_length_96_cov_1.000000-103    A:0.629098          42/97/105   \n",
       "NODE_2126217_length_62_cov_1.000000-28            T:0            244/0/0   \n",
       "NODE_2126222_length_45_cov_2.000000-70            G:1            0/0/244   \n",
       "NODE_2126233_length_45_cov_1.000000-20            G:1            0/0/244   \n",
       "NODE_2126233_length_45_cov_1.000000-45            C:1            0/0/244   \n",
       "\n",
       "                                            E(HOM1/HET/HOM2)  ChiSq_HWE  \\\n",
       "NODE_16_length_56_cov_1.375000-61           0.00/0.00/244.00        NaN   \n",
       "NODE_73_length_45_cov_1.955556-63           0.04/5.93/238.04   0.037809   \n",
       "NODE_117_length_99_cov_3.979798-22          0.00/1.00/243.00   0.001029   \n",
       "NODE_117_length_99_cov_3.979798-45          0.10/9.80/234.10   0.106791   \n",
       "NODE_117_length_99_cov_3.979798-46          0.10/9.80/234.10   0.106791   \n",
       "NODE_117_length_99_cov_3.979798-64          0.10/9.80/234.10   0.106791   \n",
       "NODE_117_length_99_cov_3.979798-72        16.27/93.47/134.27  16.809900   \n",
       "NODE_117_length_99_cov_3.979798-82         3.69/52.62/187.69   4.795179   \n",
       "NODE_117_length_99_cov_3.979798-97        11.30/82.41/150.30   4.031129   \n",
       "NODE_117_length_99_cov_3.979798-98         5.76/63.47/174.76   0.370495   \n",
       "NODE_117_length_99_cov_3.979798-102        8.86/75.28/159.86   0.003295   \n",
       "NODE_117_length_99_cov_3.979798-106       54.20/121.60/68.20  32.533620   \n",
       "NODE_117_length_99_cov_3.979798-107       61.50/122.00/60.50  46.054290   \n",
       "NODE_162_length_98_cov_5.010204-11          242.00/1.99/0.00   0.004132   \n",
       "NODE_162_length_98_cov_5.010204-13          243.00/1.00/0.00   0.001029   \n",
       "NODE_162_length_98_cov_5.010204-17         229.23/14.54/0.23   0.245386   \n",
       "NODE_162_length_98_cov_5.010204-24          243.00/1.00/0.00   0.001029   \n",
       "NODE_162_length_98_cov_5.010204-27          235.08/8.83/0.08   0.086140   \n",
       "NODE_162_length_98_cov_5.010204-28          236.07/7.87/0.07   0.067778   \n",
       "NODE_162_length_98_cov_5.010204-31         231.17/12.65/0.17   0.182763   \n",
       "NODE_162_length_98_cov_5.010204-40         208.40/34.19/1.40   1.642253   \n",
       "NODE_305_length_49_cov_1.204082-40         1.05/29.90/213.05  87.455160   \n",
       "NODE_305_length_49_cov_1.204082-82          0.00/0.00/244.00        NaN   \n",
       "NODE_411_length_91_cov_2.098901-11         202.89/39.21/1.89   2.278278   \n",
       "NODE_411_length_91_cov_2.098901-18         217.75/25.51/0.75   0.836981   \n",
       "NODE_411_length_91_cov_2.098901-21         207.48/35.04/1.48   1.739931   \n",
       "NODE_411_length_91_cov_2.098901-23         176.46/62.08/5.46   7.549868   \n",
       "NODE_411_length_91_cov_2.098901-25         184.20/55.61/4.20   0.012215   \n",
       "NODE_411_length_91_cov_2.098901-32         184.20/55.61/4.20   0.012215   \n",
       "NODE_550_length_45_cov_3.133333-44       122.66/100.68/20.66  17.966010   \n",
       "...                                                      ...        ...   \n",
       "NODE_2126184_length_92_cov_1.000000-35     204.72/37.56/1.72  12.662520   \n",
       "NODE_2126184_length_92_cov_1.000000-42     219.64/23.72/0.64  32.972130   \n",
       "NODE_2126184_length_92_cov_1.000000-44    142.55/87.90/13.55   1.503359   \n",
       "NODE_2126184_length_92_cov_1.000000-49     167.23/69.54/7.23   9.251424   \n",
       "NODE_2126184_length_92_cov_1.000000-50     219.64/23.72/0.64   3.206931   \n",
       "NODE_2126184_length_92_cov_1.000000-53     230.20/13.60/0.20   3.371058   \n",
       "NODE_2126184_length_92_cov_1.000000-65     217.75/25.51/0.75   2.355664   \n",
       "NODE_2126184_length_92_cov_1.000000-66    145.62/85.75/12.62   3.836041   \n",
       "NODE_2126184_length_92_cov_1.000000-76      235.08/8.83/0.08   0.086140   \n",
       "NODE_2126184_length_92_cov_1.000000-77     195.66/45.67/2.66  18.779550   \n",
       "NODE_2126184_length_92_cov_1.000000-80      239.03/4.95/0.03   0.026148   \n",
       "NODE_2126184_length_92_cov_1.000000-95     195.66/45.67/2.66  18.779550   \n",
       "NODE_2126204_length_90_cov_1.000000-45      0.00/0.00/244.00        NaN   \n",
       "NODE_2126204_length_90_cov_1.000000-58      0.00/0.00/244.00        NaN   \n",
       "NODE_2126204_length_90_cov_1.000000-67     0.92/28.16/214.92   1.430368   \n",
       "NODE_2126204_length_90_cov_1.000000-76   26.56/107.88/109.56   2.590849   \n",
       "NODE_2126204_length_90_cov_1.000000-77     7.23/69.54/167.23   4.592979   \n",
       "NODE_2126213_length_59_cov_1.000000-14      0.00/0.00/244.00        NaN   \n",
       "NODE_2126213_length_59_cov_1.000000-45      0.00/0.00/244.00        NaN   \n",
       "NODE_2126213_length_59_cov_1.000000-59      0.00/0.00/244.00        NaN   \n",
       "NODE_2126214_length_96_cov_1.000000-7     94.69/114.62/34.69   1.632973   \n",
       "NODE_2126214_length_96_cov_1.000000-18    132.05/94.90/17.05   0.097764   \n",
       "NODE_2126214_length_96_cov_1.000000-63     1.12/30.77/212.12   8.576221   \n",
       "NODE_2126214_length_96_cov_1.000000-92      243.00/1.00/0.00   0.001029   \n",
       "NODE_2126214_length_96_cov_1.000000-99   27.56/108.89/107.56   8.976994   \n",
       "NODE_2126214_length_96_cov_1.000000-103   33.57/113.87/96.57   5.353787   \n",
       "NODE_2126217_length_62_cov_1.000000-28      244.00/0.00/0.00        NaN   \n",
       "NODE_2126222_length_45_cov_2.000000-70      0.00/0.00/244.00        NaN   \n",
       "NODE_2126233_length_45_cov_1.000000-20      0.00/0.00/244.00        NaN   \n",
       "NODE_2126233_length_45_cov_1.000000-45      0.00/0.00/244.00        NaN   \n",
       "\n",
       "                                                P_HWE  P_HET_DEFICIT  \\\n",
       "NODE_16_length_56_cov_1.375000-61        1.000000e+00   1.000000e+00   \n",
       "NODE_73_length_45_cov_1.955556-63        1.000000e+00   1.000000e+00   \n",
       "NODE_117_length_99_cov_3.979798-22       1.000000e+00   1.000000e+00   \n",
       "NODE_117_length_99_cov_3.979798-45       1.000000e+00   1.000000e+00   \n",
       "NODE_117_length_99_cov_3.979798-46       1.000000e+00   1.000000e+00   \n",
       "NODE_117_length_99_cov_3.979798-64       1.000000e+00   1.000000e+00   \n",
       "NODE_117_length_99_cov_3.979798-72       1.834113e-05   9.999986e-01   \n",
       "NODE_117_length_99_cov_3.979798-82       3.172595e-02   1.000000e+00   \n",
       "NODE_117_length_99_cov_3.979798-97       5.683300e-02   9.887311e-01   \n",
       "NODE_117_length_99_cov_3.979798-98       4.699040e-01   3.337456e-01   \n",
       "NODE_117_length_99_cov_3.979798-102      1.000000e+00   5.363668e-01   \n",
       "NODE_117_length_99_cov_3.979798-106      1.258374e-08   1.000000e+00   \n",
       "NODE_117_length_99_cov_3.979798-107      1.311706e-11   1.000000e+00   \n",
       "NODE_162_length_98_cov_5.010204-11       1.000000e+00   1.000000e+00   \n",
       "NODE_162_length_98_cov_5.010204-13       1.000000e+00   1.000000e+00   \n",
       "NODE_162_length_98_cov_5.010204-17       1.000000e+00   1.000000e+00   \n",
       "NODE_162_length_98_cov_5.010204-24       1.000000e+00   1.000000e+00   \n",
       "NODE_162_length_98_cov_5.010204-27       1.000000e+00   1.000000e+00   \n",
       "NODE_162_length_98_cov_5.010204-28       1.000000e+00   1.000000e+00   \n",
       "NODE_162_length_98_cov_5.010204-31       1.000000e+00   1.000000e+00   \n",
       "NODE_162_length_98_cov_5.010204-40       3.741141e-01   1.000000e+00   \n",
       "NODE_305_length_49_cov_1.204082-40       2.083831e-10   2.083831e-10   \n",
       "NODE_305_length_49_cov_1.204082-82       1.000000e+00   1.000000e+00   \n",
       "NODE_411_length_91_cov_2.098901-11       2.305470e-01   1.000000e+00   \n",
       "NODE_411_length_91_cov_2.098901-18       1.000000e+00   1.000000e+00   \n",
       "NODE_411_length_91_cov_2.098901-21       3.744188e-01   1.000000e+00   \n",
       "NODE_411_length_91_cov_2.098901-23       3.858700e-03   1.000000e+00   \n",
       "NODE_411_length_91_cov_2.098901-25       1.000000e+00   6.227237e-01   \n",
       "NODE_411_length_91_cov_2.098901-32       1.000000e+00   6.227237e-01   \n",
       "NODE_550_length_45_cov_3.133333-44       1.094846e-05   9.999986e-01   \n",
       "...                                               ...            ...   \n",
       "NODE_2126184_length_92_cov_1.000000-35   3.047546e-03   3.047546e-03   \n",
       "NODE_2126184_length_92_cov_1.000000-42   9.742994e-05   9.742994e-05   \n",
       "NODE_2126184_length_92_cov_1.000000-44   2.159177e-01   1.397860e-01   \n",
       "NODE_2126184_length_92_cov_1.000000-49   5.456015e-03   3.446771e-03   \n",
       "NODE_2126184_length_92_cov_1.000000-50   1.198603e-01   1.198603e-01   \n",
       "NODE_2126184_length_92_cov_1.000000-53   1.745334e-01   1.745334e-01   \n",
       "NODE_2126184_length_92_cov_1.000000-65   1.566682e-01   1.566682e-01   \n",
       "NODE_2126184_length_92_cov_1.000000-66   6.638337e-02   3.747657e-02   \n",
       "NODE_2126184_length_92_cov_1.000000-76   1.000000e+00   1.000000e+00   \n",
       "NODE_2126184_length_92_cov_1.000000-77   2.478658e-04   2.478658e-04   \n",
       "NODE_2126184_length_92_cov_1.000000-80   1.000000e+00   1.000000e+00   \n",
       "NODE_2126184_length_92_cov_1.000000-95   2.478658e-04   2.478658e-04   \n",
       "NODE_2126204_length_90_cov_1.000000-45   1.000000e+00   1.000000e+00   \n",
       "NODE_2126204_length_90_cov_1.000000-58   1.000000e+00   1.000000e+00   \n",
       "NODE_2126204_length_90_cov_1.000000-67   2.215542e-01   2.215542e-01   \n",
       "NODE_2126204_length_90_cov_1.000000-76   1.467330e-01   9.589912e-01   \n",
       "NODE_2126204_length_90_cov_1.000000-77   4.080076e-02   2.936487e-02   \n",
       "NODE_2126213_length_59_cov_1.000000-14   1.000000e+00   1.000000e+00   \n",
       "NODE_2126213_length_59_cov_1.000000-45   1.000000e+00   1.000000e+00   \n",
       "NODE_2126213_length_59_cov_1.000000-59   1.000000e+00   1.000000e+00   \n",
       "NODE_2126214_length_96_cov_1.000000-7    2.226742e-01   9.172932e-01   \n",
       "NODE_2126214_length_96_cov_1.000000-18   7.431973e-01   4.237454e-01   \n",
       "NODE_2126214_length_96_cov_1.000000-63   1.585400e-02   1.585400e-02   \n",
       "NODE_2126214_length_96_cov_1.000000-92   1.000000e+00   1.000000e+00   \n",
       "NODE_2126214_length_96_cov_1.000000-99   3.872338e-03   2.083067e-03   \n",
       "NODE_2126214_length_96_cov_1.000000-103  2.028657e-02   1.375216e-02   \n",
       "NODE_2126217_length_62_cov_1.000000-28   1.000000e+00   1.000000e+00   \n",
       "NODE_2126222_length_45_cov_2.000000-70   1.000000e+00   1.000000e+00   \n",
       "NODE_2126233_length_45_cov_1.000000-20   1.000000e+00   1.000000e+00   \n",
       "NODE_2126233_length_45_cov_1.000000-45   1.000000e+00   1.000000e+00   \n",
       "\n",
       "                                         P_HET_EXCESS A1_allele A2_allele  \\\n",
       "NODE_16_length_56_cov_1.375000-61        1.000000e+00         A         T   \n",
       "NODE_73_length_45_cov_1.955556-63        9.693896e-01         T         C   \n",
       "NODE_117_length_99_cov_3.979798-22       1.000000e+00         T         C   \n",
       "NODE_117_length_99_cov_3.979798-45       9.102373e-01         T         A   \n",
       "NODE_117_length_99_cov_3.979798-46       9.102373e-01         T         A   \n",
       "NODE_117_length_99_cov_3.979798-64       9.102373e-01         T         G   \n",
       "NODE_117_length_99_cov_3.979798-72       1.048217e-05         T         C   \n",
       "NODE_117_length_99_cov_3.979798-82       1.585835e-02         A         G   \n",
       "NODE_117_length_99_cov_3.979798-97       3.198435e-02         T         C   \n",
       "NODE_117_length_99_cov_3.979798-98       8.160896e-01         A         C   \n",
       "NODE_117_length_99_cov_3.979798-102      6.263141e-01         T         C   \n",
       "NODE_117_length_99_cov_3.979798-106      9.825594e-09         T         C   \n",
       "NODE_117_length_99_cov_3.979798-107      8.048139e-12         A         G   \n",
       "NODE_162_length_98_cov_5.010204-11       9.979466e-01         T         C   \n",
       "NODE_162_length_98_cov_5.010204-13       1.000000e+00         C         T   \n",
       "NODE_162_length_98_cov_5.010204-17       8.010857e-01         G         A   \n",
       "NODE_162_length_98_cov_5.010204-24       1.000000e+00         C         T   \n",
       "NODE_162_length_98_cov_5.010204-27       9.276674e-01         C         T   \n",
       "NODE_162_length_98_cov_5.010204-28       9.433905e-01         G         A   \n",
       "NODE_162_length_98_cov_5.010204-31       8.486940e-01         C         A   \n",
       "NODE_162_length_98_cov_5.010204-40       2.283920e-01         G         A   \n",
       "NODE_305_length_49_cov_1.204082-40       1.000000e+00         A         T   \n",
       "NODE_305_length_49_cov_1.204082-82       1.000000e+00         A         C   \n",
       "NODE_411_length_91_cov_2.098901-11       1.313331e-01         C         T   \n",
       "NODE_411_length_91_cov_2.098901-18       4.672079e-01         T         G   \n",
       "NODE_411_length_91_cov_2.098901-21       2.096547e-01         T         C   \n",
       "NODE_411_length_91_cov_2.098901-23       1.733884e-03         G         T   \n",
       "NODE_411_length_91_cov_2.098901-25       6.000197e-01         C         T   \n",
       "NODE_411_length_91_cov_2.098901-32       6.000197e-01         C         G   \n",
       "NODE_550_length_45_cov_3.133333-44       8.184367e-06         C         T   \n",
       "...                                               ...       ...       ...   \n",
       "NODE_2126184_length_92_cov_1.000000-35   9.995941e-01         C         T   \n",
       "NODE_2126184_length_92_cov_1.000000-42   9.999963e-01         T         C   \n",
       "NODE_2126184_length_92_cov_1.000000-44   9.229479e-01         T         G   \n",
       "NODE_2126184_length_92_cov_1.000000-49   9.990364e-01         A         C   \n",
       "NODE_2126184_length_92_cov_1.000000-50   9.822594e-01         T         C   \n",
       "NODE_2126184_length_92_cov_1.000000-53   9.880585e-01         G         T   \n",
       "NODE_2126184_length_92_cov_1.000000-65   9.721413e-01         A         G   \n",
       "NODE_2126184_length_92_cov_1.000000-66   9.832893e-01         G         A   \n",
       "NODE_2126184_length_92_cov_1.000000-76   9.276674e-01         G         T   \n",
       "NODE_2126184_length_92_cov_1.000000-77   9.999687e-01         C         G   \n",
       "NODE_2126184_length_92_cov_1.000000-80   9.795296e-01         T         C   \n",
       "NODE_2126184_length_92_cov_1.000000-95   9.999687e-01         G         A   \n",
       "NODE_2126204_length_90_cov_1.000000-45   1.000000e+00         T         C   \n",
       "NODE_2126204_length_90_cov_1.000000-58   1.000000e+00         C         G   \n",
       "NODE_2126204_length_90_cov_1.000000-67   9.497151e-01         T         C   \n",
       "NODE_2126204_length_90_cov_1.000000-76   7.461743e-02         G         A   \n",
       "NODE_2126204_length_90_cov_1.000000-77   9.892343e-01         A         G   \n",
       "NODE_2126213_length_59_cov_1.000000-14   1.000000e+00         A         C   \n",
       "NODE_2126213_length_59_cov_1.000000-45   1.000000e+00         T         C   \n",
       "NODE_2126213_length_59_cov_1.000000-59   1.000000e+00         T         C   \n",
       "NODE_2126214_length_96_cov_1.000000-7    1.331263e-01         C         T   \n",
       "NODE_2126214_length_96_cov_1.000000-18   6.981939e-01         T         G   \n",
       "NODE_2126214_length_96_cov_1.000000-63   9.978919e-01         T         A   \n",
       "NODE_2126214_length_96_cov_1.000000-92   1.000000e+00         T         C   \n",
       "NODE_2126214_length_96_cov_1.000000-99   9.991796e-01         A         G   \n",
       "NODE_2126214_length_96_cov_1.000000-103  9.933807e-01         G         A   \n",
       "NODE_2126217_length_62_cov_1.000000-28   1.000000e+00         A         T   \n",
       "NODE_2126222_length_45_cov_2.000000-70   1.000000e+00         A         G   \n",
       "NODE_2126233_length_45_cov_1.000000-20   1.000000e+00         A         G   \n",
       "NODE_2126233_length_45_cov_1.000000-45   1.000000e+00         G         C   \n",
       "\n",
       "                                          A1_freq   A2_freq       MAF  \\\n",
       "NODE_16_length_56_cov_1.375000-61        0.000000  1.000000  0.000000   \n",
       "NODE_73_length_45_cov_1.955556-63        0.012295  0.987705  0.012295   \n",
       "NODE_117_length_99_cov_3.979798-22       0.002049  0.997951  0.002049   \n",
       "NODE_117_length_99_cov_3.979798-45       0.020492  0.979508  0.020492   \n",
       "NODE_117_length_99_cov_3.979798-46       0.020492  0.979508  0.020492   \n",
       "NODE_117_length_99_cov_3.979798-64       0.020492  0.979508  0.020492   \n",
       "NODE_117_length_99_cov_3.979798-72       0.258197  0.741803  0.258197   \n",
       "NODE_117_length_99_cov_3.979798-82       0.122951  0.877049  0.122951   \n",
       "NODE_117_length_99_cov_3.979798-97       0.215164  0.784836  0.215164   \n",
       "NODE_117_length_99_cov_3.979798-98       0.153689  0.846311  0.153689   \n",
       "NODE_117_length_99_cov_3.979798-102      0.190574  0.809426  0.190574   \n",
       "NODE_117_length_99_cov_3.979798-106      0.471311  0.528689  0.471311   \n",
       "NODE_117_length_99_cov_3.979798-107      0.502049  0.497951  0.497951   \n",
       "NODE_162_length_98_cov_5.010204-11       0.995902  0.004098  0.004098   \n",
       "NODE_162_length_98_cov_5.010204-13       0.997951  0.002049  0.002049   \n",
       "NODE_162_length_98_cov_5.010204-17       0.969262  0.030738  0.030738   \n",
       "NODE_162_length_98_cov_5.010204-24       0.997951  0.002049  0.002049   \n",
       "NODE_162_length_98_cov_5.010204-27       0.981557  0.018443  0.018443   \n",
       "NODE_162_length_98_cov_5.010204-28       0.983607  0.016393  0.016393   \n",
       "NODE_162_length_98_cov_5.010204-31       0.973361  0.026639  0.026639   \n",
       "NODE_162_length_98_cov_5.010204-40       0.924180  0.075820  0.075820   \n",
       "NODE_305_length_49_cov_1.204082-40       0.065574  0.934426  0.065574   \n",
       "NODE_305_length_49_cov_1.204082-82       0.000000  1.000000  0.000000   \n",
       "NODE_411_length_91_cov_2.098901-11       0.911885  0.088115  0.088115   \n",
       "NODE_411_length_91_cov_2.098901-18       0.944672  0.055328  0.055328   \n",
       "NODE_411_length_91_cov_2.098901-21       0.922131  0.077869  0.077869   \n",
       "NODE_411_length_91_cov_2.098901-23       0.850410  0.149590  0.149590   \n",
       "NODE_411_length_91_cov_2.098901-25       0.868852  0.131148  0.131148   \n",
       "NODE_411_length_91_cov_2.098901-32       0.868852  0.131148  0.131148   \n",
       "NODE_550_length_45_cov_3.133333-44       0.709016  0.290984  0.290984   \n",
       "...                                           ...       ...       ...   \n",
       "NODE_2126184_length_92_cov_1.000000-35   0.915984  0.084016  0.084016   \n",
       "NODE_2126184_length_92_cov_1.000000-42   0.948770  0.051229  0.051229   \n",
       "NODE_2126184_length_92_cov_1.000000-44   0.764344  0.235656  0.235656   \n",
       "NODE_2126184_length_92_cov_1.000000-49   0.827869  0.172131  0.172131   \n",
       "NODE_2126184_length_92_cov_1.000000-50   0.948770  0.051229  0.051229   \n",
       "NODE_2126184_length_92_cov_1.000000-53   0.971311  0.028688  0.028688   \n",
       "NODE_2126184_length_92_cov_1.000000-65   0.944672  0.055328  0.055328   \n",
       "NODE_2126184_length_92_cov_1.000000-66   0.772541  0.227459  0.227459   \n",
       "NODE_2126184_length_92_cov_1.000000-76   0.981557  0.018443  0.018443   \n",
       "NODE_2126184_length_92_cov_1.000000-77   0.895492  0.104508  0.104508   \n",
       "NODE_2126184_length_92_cov_1.000000-80   0.989754  0.010246  0.010246   \n",
       "NODE_2126184_length_92_cov_1.000000-95   0.895492  0.104508  0.104508   \n",
       "NODE_2126204_length_90_cov_1.000000-45   0.000000  1.000000  0.000000   \n",
       "NODE_2126204_length_90_cov_1.000000-58   0.000000  1.000000  0.000000   \n",
       "NODE_2126204_length_90_cov_1.000000-67   0.061475  0.938525  0.061475   \n",
       "NODE_2126204_length_90_cov_1.000000-76   0.329918  0.670082  0.329918   \n",
       "NODE_2126204_length_90_cov_1.000000-77   0.172131  0.827869  0.172131   \n",
       "NODE_2126213_length_59_cov_1.000000-14   0.000000  1.000000  0.000000   \n",
       "NODE_2126213_length_59_cov_1.000000-45   0.000000  1.000000  0.000000   \n",
       "NODE_2126213_length_59_cov_1.000000-59   0.000000  1.000000  0.000000   \n",
       "NODE_2126214_length_96_cov_1.000000-7    0.622951  0.377049  0.377049   \n",
       "NODE_2126214_length_96_cov_1.000000-18   0.735656  0.264344  0.264344   \n",
       "NODE_2126214_length_96_cov_1.000000-63   0.067623  0.932377  0.067623   \n",
       "NODE_2126214_length_96_cov_1.000000-92   0.997951  0.002049  0.002049   \n",
       "NODE_2126214_length_96_cov_1.000000-99   0.336066  0.663934  0.336066   \n",
       "NODE_2126214_length_96_cov_1.000000-103  0.370902  0.629098  0.370902   \n",
       "NODE_2126217_length_62_cov_1.000000-28   1.000000  0.000000  0.000000   \n",
       "NODE_2126222_length_45_cov_2.000000-70   0.000000  1.000000  0.000000   \n",
       "NODE_2126233_length_45_cov_1.000000-20   0.000000  1.000000  0.000000   \n",
       "NODE_2126233_length_45_cov_1.000000-45   0.000000  1.000000  0.000000   \n",
       "\n",
       "                                              Fis  \n",
       "NODE_16_length_56_cov_1.375000-61             NaN  \n",
       "NODE_73_length_45_cov_1.955556-63       -0.010373  \n",
       "NODE_117_length_99_cov_3.979798-22       0.000000  \n",
       "NODE_117_length_99_cov_3.979798-45      -0.018828  \n",
       "NODE_117_length_99_cov_3.979798-46      -0.018828  \n",
       "NODE_117_length_99_cov_3.979798-64      -0.018828  \n",
       "NODE_117_length_99_cov_3.979798-72      -0.259888  \n",
       "NODE_117_length_99_cov_3.979798-82      -0.137850  \n",
       "NODE_117_length_99_cov_3.979798-97      -0.126222  \n",
       "NODE_117_length_99_cov_3.979798-98       0.040936  \n",
       "NODE_117_length_99_cov_3.979798-102      0.005717  \n",
       "NODE_117_length_99_cov_3.979798-106     -0.362353  \n",
       "NODE_117_length_99_cov_3.979798-107     -0.431511  \n",
       "NODE_162_length_98_cov_5.010204-11      -0.002058  \n",
       "NODE_162_length_98_cov_5.010204-13       0.000000  \n",
       "NODE_162_length_98_cov_5.010204-17      -0.029598  \n",
       "NODE_162_length_98_cov_5.010204-24       0.000000  \n",
       "NODE_162_length_98_cov_5.010204-27      -0.016701  \n",
       "NODE_162_length_98_cov_5.010204-28      -0.014583  \n",
       "NODE_162_length_98_cov_5.010204-31      -0.025263  \n",
       "NODE_162_length_98_cov_5.010204-40      -0.079823  \n",
       "NODE_305_length_49_cov_1.204082-40       0.599507  \n",
       "NODE_305_length_49_cov_1.204082-82            NaN  \n",
       "NODE_411_length_91_cov_2.098901-11      -0.094382  \n",
       "NODE_411_length_91_cov_2.098901-18      -0.056399  \n",
       "NODE_411_length_91_cov_2.098901-21      -0.082222  \n",
       "NODE_411_length_91_cov_2.098901-23      -0.173494  \n",
       "NODE_411_length_91_cov_2.098901-25      -0.005012  \n",
       "NODE_411_length_91_cov_2.098901-32      -0.005012  \n",
       "NODE_550_length_45_cov_3.133333-44      -0.268745  \n",
       "...                                           ...  \n",
       "NODE_2126184_length_92_cov_1.000000-35   0.229388  \n",
       "NODE_2126184_length_92_cov_1.000000-42   0.368898  \n",
       "NODE_2126184_length_92_cov_1.000000-44   0.080382  \n",
       "NODE_2126184_length_92_cov_1.000000-49   0.196370  \n",
       "NODE_2126184_length_92_cov_1.000000-50   0.116458  \n",
       "NODE_2126184_length_92_cov_1.000000-53   0.119349  \n",
       "NODE_2126184_length_92_cov_1.000000-65   0.100104  \n",
       "NODE_2126184_length_92_cov_1.000000-66   0.127178  \n",
       "NODE_2126184_length_92_cov_1.000000-76  -0.016701  \n",
       "NODE_2126184_length_92_cov_1.000000-77   0.278907  \n",
       "NODE_2126184_length_92_cov_1.000000-80  -0.008282  \n",
       "NODE_2126184_length_92_cov_1.000000-95   0.278907  \n",
       "NODE_2126204_length_90_cov_1.000000-45        NaN  \n",
       "NODE_2126204_length_90_cov_1.000000-58        NaN  \n",
       "NODE_2126204_length_90_cov_1.000000-67   0.078457  \n",
       "NODE_2126204_length_90_cov_1.000000-76  -0.100784  \n",
       "NODE_2126204_length_90_cov_1.000000-77   0.138967  \n",
       "NODE_2126213_length_59_cov_1.000000-14        NaN  \n",
       "NODE_2126213_length_59_cov_1.000000-45        NaN  \n",
       "NODE_2126213_length_59_cov_1.000000-59        NaN  \n",
       "NODE_2126214_length_96_cov_1.000000-7   -0.079591  \n",
       "NODE_2126214_length_96_cov_1.000000-18   0.022025  \n",
       "NODE_2126214_length_96_cov_1.000000-63   0.189144  \n",
       "NODE_2126214_length_96_cov_1.000000-92   0.000000  \n",
       "NODE_2126214_length_96_cov_1.000000-99   0.193466  \n",
       "NODE_2126214_length_96_cov_1.000000-103  0.149873  \n",
       "NODE_2126217_length_62_cov_1.000000-28        NaN  \n",
       "NODE_2126222_length_45_cov_2.000000-70        NaN  \n",
       "NODE_2126233_length_45_cov_1.000000-20        NaN  \n",
       "NODE_2126233_length_45_cov_1.000000-45        NaN  \n",
       "\n",
       "[1300961 rows x 25 columns]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df_beagle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>N_CHR</th>\n",
       "      <th>A1_FREQ</th>\n",
       "      <th>A2_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE_16_length_56_cov_1.375000-61</th>\n",
       "      <td>NODE_16_length_56_cov_1.375000</td>\n",
       "      <td>61</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0</td>\n",
       "      <td>T:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_73_length_45_cov_1.955556-63</th>\n",
       "      <td>NODE_73_length_45_cov_1.955556</td>\n",
       "      <td>63</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.0122951</td>\n",
       "      <td>C:0.987705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-22</th>\n",
       "      <td>NODE_117_length_99_cov_3.979798</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.00204918</td>\n",
       "      <td>C:0.997951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-45</th>\n",
       "      <td>NODE_117_length_99_cov_3.979798</td>\n",
       "      <td>45</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.0204918</td>\n",
       "      <td>A:0.979508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-46</th>\n",
       "      <td>NODE_117_length_99_cov_3.979798</td>\n",
       "      <td>46</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.0204918</td>\n",
       "      <td>A:0.979508</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                              CHROM POS  \\\n",
       "NODE_16_length_56_cov_1.375000-61    NODE_16_length_56_cov_1.375000  61   \n",
       "NODE_73_length_45_cov_1.955556-63    NODE_73_length_45_cov_1.955556  63   \n",
       "NODE_117_length_99_cov_3.979798-22  NODE_117_length_99_cov_3.979798  22   \n",
       "NODE_117_length_99_cov_3.979798-45  NODE_117_length_99_cov_3.979798  45   \n",
       "NODE_117_length_99_cov_3.979798-46  NODE_117_length_99_cov_3.979798  46   \n",
       "\n",
       "                                   N_ALLELES N_CHR       A1_FREQ     A2_FREQ  \n",
       "NODE_16_length_56_cov_1.375000-61          2   488           A:0         T:1  \n",
       "NODE_73_length_45_cov_1.955556-63          2   488   T:0.0122951  C:0.987705  \n",
       "NODE_117_length_99_cov_3.979798-22         2   488  T:0.00204918  C:0.997951  \n",
       "NODE_117_length_99_cov_3.979798-45         2   488   T:0.0204918  A:0.979508  \n",
       "NODE_117_length_99_cov_3.979798-46         2   488   T:0.0204918  A:0.979508  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "freq_df_beagle.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "      <th>E(HOM1/HET/HOM2)</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE_16_length_56_cov_1.375000-61</th>\n",
       "      <td>NODE_16_length_56_cov_1.375000</td>\n",
       "      <td>61</td>\n",
       "      <td>0/0/244</td>\n",
       "      <td>0.00/0.00/244.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_73_length_45_cov_1.955556-63</th>\n",
       "      <td>NODE_73_length_45_cov_1.955556</td>\n",
       "      <td>63</td>\n",
       "      <td>0/6/238</td>\n",
       "      <td>0.04/5.93/238.04</td>\n",
       "      <td>0.037809</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.969390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-22</th>\n",
       "      <td>NODE_117_length_99_cov_3.979798</td>\n",
       "      <td>22</td>\n",
       "      <td>0/1/243</td>\n",
       "      <td>0.00/1.00/243.00</td>\n",
       "      <td>0.001029</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-45</th>\n",
       "      <td>NODE_117_length_99_cov_3.979798</td>\n",
       "      <td>45</td>\n",
       "      <td>0/10/234</td>\n",
       "      <td>0.10/9.80/234.10</td>\n",
       "      <td>0.106791</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.910237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_117_length_99_cov_3.979798-46</th>\n",
       "      <td>NODE_117_length_99_cov_3.979798</td>\n",
       "      <td>46</td>\n",
       "      <td>0/10/234</td>\n",
       "      <td>0.10/9.80/234.10</td>\n",
       "      <td>0.106791</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.910237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                              CHROM  POS  \\\n",
       "NODE_16_length_56_cov_1.375000-61    NODE_16_length_56_cov_1.375000   61   \n",
       "NODE_73_length_45_cov_1.955556-63    NODE_73_length_45_cov_1.955556   63   \n",
       "NODE_117_length_99_cov_3.979798-22  NODE_117_length_99_cov_3.979798   22   \n",
       "NODE_117_length_99_cov_3.979798-45  NODE_117_length_99_cov_3.979798   45   \n",
       "NODE_117_length_99_cov_3.979798-46  NODE_117_length_99_cov_3.979798   46   \n",
       "\n",
       "                                   OBS(HOM1/HET/HOM2)  E(HOM1/HET/HOM2)  \\\n",
       "NODE_16_length_56_cov_1.375000-61             0/0/244  0.00/0.00/244.00   \n",
       "NODE_73_length_45_cov_1.955556-63             0/6/238  0.04/5.93/238.04   \n",
       "NODE_117_length_99_cov_3.979798-22            0/1/243  0.00/1.00/243.00   \n",
       "NODE_117_length_99_cov_3.979798-45           0/10/234  0.10/9.80/234.10   \n",
       "NODE_117_length_99_cov_3.979798-46           0/10/234  0.10/9.80/234.10   \n",
       "\n",
       "                                    ChiSq_HWE  P_HWE  P_HET_DEFICIT  \\\n",
       "NODE_16_length_56_cov_1.375000-61         NaN      1              1   \n",
       "NODE_73_length_45_cov_1.955556-63    0.037809      1              1   \n",
       "NODE_117_length_99_cov_3.979798-22   0.001029      1              1   \n",
       "NODE_117_length_99_cov_3.979798-45   0.106791      1              1   \n",
       "NODE_117_length_99_cov_3.979798-46   0.106791      1              1   \n",
       "\n",
       "                                    P_HET_EXCESS  \n",
       "NODE_16_length_56_cov_1.375000-61       1.000000  \n",
       "NODE_73_length_45_cov_1.955556-63       0.969390  \n",
       "NODE_117_length_99_cov_3.979798-22      1.000000  \n",
       "NODE_117_length_99_cov_3.979798-45      0.910237  \n",
       "NODE_117_length_99_cov_3.979798-46      0.910237  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hardy_beagle.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "loci_beagle_file = os.path.join(beagle_dir,'loci_beagle_vcf_stats.txt')\n",
    "loci_df_beagle.to_csv(loci_beagle_file,header=True,index=True,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "freq_beagle = os.path.join(beagle_dir,'frq_beagle_vcf_stats.txt')\n",
    "freq_df_beagle.to_csv(freq_beagle,header=True,index=True,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hardy_beagle_file = os.path.join(beagle_dir,'hwe_beagle_vcf_stats.txt')\n",
    "hardy_beagle.to_csv(hardy_beagle_file,header=True,index=True,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "chroms = sorted(set([x.split(\"-\")[0] for x in loci_df.index]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/lindb/wbp/concatenated/snps/'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "with open(os.path.join(snp_dir, \"chrom_map.txt\"), \"w\") as o:\n",
    "    for i, c in enumerate(chroms):\n",
    "        o.write(\"%s\\t%d\\n\" % (c, i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def write_plink_files(vcf_gz):\n",
    "    !$vcftools --gzvcf {vcf_gz} \\\n",
    "    --out {vcf_gz} \\\n",
    "    --plink \\\n",
    "    --chrom-map {os.path.join(snp_dir, \"chrom_map.txt\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz'"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_filtered_gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--chrom-map /home/lindb/wbp/concatenated/snps/chrom_map.txt\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--plink\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Writing PLINK PED and MAP files ... \n",
      "\tRead 178750 chromosome mapping file entries.\n",
      "Done.\n",
      "After filtering, kept 1300961 out of a possible 1300961 Sites\n",
      "Run Time = 346.00 seconds\n"
     ]
    }
   ],
   "source": [
    "write_plink_files(vcf_filtered_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def write_plink_recode(vcf_gz):\n",
    "    !$plink --recodeA --tab --file {vcf_gz} --out {vcf_gz}_recodeA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "@----------------------------------------------------------@\n",
      "|        PLINK!       |     v1.07      |   10/Aug/2009     |\n",
      "|----------------------------------------------------------|\n",
      "|  (C) 2009 Shaun Purcell, GNU General Public License, v2  |\n",
      "|----------------------------------------------------------|\n",
      "|  For documentation, citation & bug-report instructions:  |\n",
      "|        http://pngu.mgh.harvard.edu/purcell/plink/        |\n",
      "@----------------------------------------------------------@\n",
      "\n",
      "Skipping web check... [ --noweb ] \n",
      "Writing this text to log file [ /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz_recodeA.log ]\n",
      "Analysis started: Tue Feb 16 12:11:59 2016\n",
      "\n",
      "Options in effect:\n",
      "\t--noweb\n",
      "\t--recodeA\n",
      "\t--tab\n",
      "\t--file /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--out /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz_recodeA\n",
      "\n",
      "** For gPLINK compatibility, do not use '.' in --out **\n",
      "1300961 (of 1300961) markers to be included from [ /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz.map ]\n",
      "Warning, found 244 individuals with ambiguous sex codes\n",
      "Writing list of these individuals to [ /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz_recodeA.nosex ]\n",
      "244 individuals read from [ /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz.ped ] \n",
      "0 individuals with nonmissing phenotypes\n",
      "Assuming a disease phenotype (1=unaff, 2=aff, 0=miss)\n",
      "Missing phenotype value is also -9\n",
      "0 cases, 0 controls and 244 missing\n",
      "0 males, 0 females, and 244 of unspecified sex\n",
      "Before frequency and genotyping pruning, there are 1300961 SNPs\n",
      "244 founders and 0 non-founders found\n",
      "87 heterozygous haploid genotypes; set to missing\n",
      "Writing list of heterozygous haploid genotypes to [ /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz_recodeA.hh ]\n",
      "Total genotyping rate in remaining individuals is nan\n",
      "0 SNPs failed missingness test ( GENO > 1 )\n",
      "0 SNPs failed frequency test ( MAF < 0 )\n",
      "After frequency and genotyping pruning, there are 1300961 SNPs\n",
      "After filtering, 0 cases, 0 controls and 244 missing\n",
      "After filtering, 0 males, 0 females, and 244 of unspecified sex\n",
      "Writing recoded file to [ /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz_recodeA.raw ] \n",
      "\n",
      "Analysis finished: Tue Feb 16 12:17:40 2016\n",
      "\n"
     ]
    }
   ],
   "source": [
    "write_plink_recode(vcf_filtered_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    1300961.000000\n",
       "mean        1745.176835\n",
       "std         3104.671683\n",
       "min          139.000000\n",
       "25%          535.000000\n",
       "50%          870.000000\n",
       "75%         1671.000000\n",
       "max        56722.000000\n",
       "Name: SUM_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.SUM_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x2acb3bdaedd0>"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAaMAAAEACAYAAAAeHRm0AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAF9tJREFUeJzt3X3wXNV93/H3B8kYsJEMTRANAozLg43HjUOKaOt6ohCD\nwB4L2hlcMqkRhmQ6Ax27ydQ1eDIgxtOx42kanLZ40poYoSQlGD8gt9QIBvA0HhzAhkIsItRxwEhY\nP1IeRI09gMS3f+yRtZL1sBK/1Vm079fMju/97j13z73+sR/ds2fvpqqQJKmng3p3QJIkw0iS1J1h\nJEnqzjCSJHVnGEmSujOMJEndjT2Mkjye5H8neTDJfa12RJLVSdYmuT3J/KHtr0yyLsmjSc4eqp+W\n5OEkjyW5dqh+cJKbWpt7kxw39Nyytv3aJBeN+1glSftmf1wZvQosrqpfqqpFrXYFcGdVnQLcBVwJ\nkORU4EPAO4BzgeuSpLX5PHBpVZ0MnJxkSatfCjxbVScB1wKfbfs6ArgKOB04A7h6OPQkSZNjf4RR\ndvI65wEr2vIK4Py2vBS4qao2V9XjwDpgUZKjgcOr6v623Y1DbYb3dQtwZlteAqyuqk1V9TywGjhn\n1o5KkjRr9kcYFXBHkvuT/GarLaiqGYCq2ggc1erHAE8Otd3QascA64fq61ttuzZVtQXYlOTI3exL\nkjRh5u6H13hPVf0wyc8Dq5OsZRBQw2bznkTZ8yaSpEky9jCqqh+2//3bJF8DFgEzSRZU1Uwbgnu6\nbb4BOHao+cJW21V9uM1TSeYA86rq2SQbgMU7tLl7x/4l8eZ8krQPqmrW/vE/1mG6JIcleXNbfhNw\nNvAIsAq4uG22DLi1La8CLmwz5E4ATgTua0N5m5IsahMaLtqhzbK2fAGDCREAtwNnJZnfJjOc1Wo/\no6p8VHH11Vd378OkPDwXngvPxe4fs23cV0YLgK+2q4+5wJ9W1eokDwA3J7kEeILBDDqqak2Sm4E1\nwCvAZbXtqC8HbgAOAW6rqm+0+vXAyiTrgGeAC9u+nkvyKeABBsOA19RgIoMkacKMNYyq6m+Ad++k\n/izwvl20+TTw6Z3UvwO8ayf1l2hhtpPnbmAQYJKkCeYdGPRTixcv7t2FieG52MZzsY3nYnwyjrG/\n15MkNe3nQJL2VhLq9TKBQZKkURhGkqTuDCNJUneGkSSpO8NIktSdYSRJ6s4wkiR1ZxhJkrozjCRJ\n3RlGkqTuDCNJUneGkSSpu/3xs+OSpFl09NFvZWbmid7dmFXetdu7dkt6nRn84HXv9y3v2i1JOsAY\nRpKk7gwjSVJ3hpEkqTvDSJLUnWEkSerOMJIkdWcYSZK6M4wkSd0ZRpKk7gwjSVJ3hpEkqTvDSJLU\nnWEkSerOMJIkdWcYSZK6M4wkSd0ZRpKk7gwjSVJ3hpEkqbv9EkZJDkry3SSr2voRSVYnWZvk9iTz\nh7a9Msm6JI8mOXuoflqSh5M8luTaofrBSW5qbe5NctzQc8va9muTXLQ/jlWStPf215XRx4A1Q+tX\nAHdW1SnAXcCVAElOBT4EvAM4F7guSVqbzwOXVtXJwMlJlrT6pcCzVXUScC3w2bavI4CrgNOBM4Cr\nh0NPkjQ5xh5GSRYC7we+MFQ+D1jRllcA57flpcBNVbW5qh4H1gGLkhwNHF5V97ftbhxqM7yvW4Az\n2/ISYHVVbaqq54HVwDmzeWySpNmxP66M/gD4OFBDtQVVNQNQVRuBo1r9GODJoe02tNoxwPqh+vpW\n265NVW0BNiU5cjf7kiRNmLnj3HmSDwAzVfVQksW72bR289xev+zeNli+fPlPlxcvXszixYtnsTuS\ndCC4pz3GY6xhBLwHWJrk/cChwOFJVgIbkyyoqpk2BPd0234DcOxQ+4Wttqv6cJunkswB5lXVs0k2\nAIt3aHP3zjo5HEaSpJ1ZzPZvqdfM6t7HOkxXVZ+squOq6m3AhcBdVfVh4OvAxW2zZcCtbXkVcGGb\nIXcCcCJwXxvK25RkUZvQcNEObZa15QsYTIgAuB04K8n8NpnhrFaTJE2YcV8Z7cpngJuTXAI8wWAG\nHVW1JsnNDGbevQJcVlVbh/AuB24ADgFuq6pvtPr1wMok64BnGIQeVfVckk8BDzAYBrymTWSQJE2Y\nbHuvn05JatrPgaTXl8EAUe/3rVBVe/0Z/a54BwZJUneGkSSpO8NIktSdYSRJ6s4wkiR1ZxhJkroz\njCRJ3RlGkqTuDCNJUneGkSSpO8NIktSdYSRJ6q7XXbsnyksvvcQrr7zStQ8HHXQQhx12WNc+SFIv\n3rU7qfnzf54f//jHXfvx6qsv861v/S/OOOOMrv2QNPkOxLt2e2UEvPjii2ze/GLXPsybt5SZmZmu\nfZCkXvzMSJLUnWEkSerOMJIkdWcYSZK6M4wkSd0ZRpKk7gwjSVJ3hpEkqTvDSJLUnWEkSerOMJIk\ndWcYSZK6M4wkSd0ZRpKk7gwjSVJ3hpEkqTvDSJLUnWEkSerOMJIkdWcYSZK6G2sYJXljkr9M8mCS\nR5Jc3epHJFmdZG2S25PMH2pzZZJ1SR5NcvZQ/bQkDyd5LMm1Q/WDk9zU2tyb5Lih55a17dcmuWic\nxypJ2ndjDaOqegn41ar6JeDdwLlJFgFXAHdW1SnAXcCVAElOBT4EvAM4F7guSdruPg9cWlUnAycn\nWdLqlwLPVtVJwLXAZ9u+jgCuAk4HzgCuHg49SdLkGPswXVX9uC2+EZgLFHAesKLVVwDnt+WlwE1V\ntbmqHgfWAYuSHA0cXlX3t+1uHGozvK9bgDPb8hJgdVVtqqrngdXAObN8eJKkWTBSGCV5176+QJKD\nkjwIbATuaIGyoKpmAKpqI3BU2/wY4Mmh5hta7Rhg/VB9fatt16aqtgCbkhy5m31JkibMqFdG1yW5\nL8llezvUVVWvtmG6hQyuct7J4Opou832Zp97kD1vIkmaJHNH2aiq3pvkJOAS4DtJ7gO+WFV3jPpC\nVfVCknsYDJXNJFlQVTNtCO7pttkG4NihZgtbbVf14TZPJZkDzKuqZ5NsABbv0ObunfVty5aXgeVt\nbfEOzSRJcE97jMfInxlV1Trgd4FPAL8C/GGSv07yz3bVJsnPbb2SSnIocBbwKLAKuLhttgy4tS2v\nAi5sM+ROAE4E7mtDeZuSLGoTGi7aoc2ytnwBgwkRALcDZyWZ3yYznNVqP2POnIMZhNFyDCJJ2pnF\nbHufXD7rex/pyijJ3wc+AnwAuAP4YFV9N8kvAPcCX9lF078LrEhyEIPg+/Oqui3Jt4Gbk1wCPMFg\nBh1VtSbJzcAa4BXgsqraOoR3OXADcAhwW1V9o9WvB1YmWQc8A1zY9vVckk8BDzAYBrymTWSQJE2Y\nbHuv381GyTeBLwC3VNVPdnjuw1W1ckz9G7skNXfuYWze/GLXfsybt5SVK3+TpUuXdu2HpMk3GCCa\nzY/a96kXVNWsfUY/0pURgyuin7TZarQrnUOq6sev5yCSJE2GUT8zuhM4dGj9sFaTJOk1GzWMDqmq\nH21dacuHjadLkqRpM2oYvZjktK0rSX4Z+MlutpckaWSjfmb0r4EvJXmKwZdKjwb++dh6JUmaKqN+\n6fX+JG8HTmmltVX1yvi6JUmaJqNeGcHg7tdvbW1OS0JV3TiWXkmSpsqoX3pdCfw94CFgSysXg7tn\nS5L0mox6ZfQPgFNrlG/ISpK0l0adTfdXDCYtSJI060a9Mvo5YE27W/dLW4tV5b1rJEmv2ahhtHyc\nnZAkTbdRp3Z/M8nxwElVdWeSw4A54+2aJGlajPqz478F3AL8USsdA3xtXJ2SJE2XUScwXA68B3gB\nfvpDe0eNq1OSpOkyahi9VFUvb11JMpf+P6YhSTpAjBpG30zySeDQJGcBXwK+Pr5uSZKmyahhdAXw\nt8AjwL8EbgN+d1ydkiRNl1Fn070K/Nf2kCRpVo16b7q/YSefEVXV22a9R5KkqbM396bb6hDgAuDI\n2e+OJGkajfSZUVU9M/TYUFXXAh8Yc98kSVNi1GG604ZWD2JwpbQ3v4UkSdIujRoovz+0vBl4HPjQ\nrPdGkjSVRp1N96vj7ogkaXqNOkz3O7t7vqr+w+x0R5I0jfZmNt3pwKq2/kHgPmDdODolSZouo4bR\nQuC0qvp/AEmWA/+jqv7FuDomSZoeo94OaAHw8tD6y60mSdJrNuqV0Y3AfUm+2tbPB1aMp0uSpGkz\n6my6f5fkfwLvbaWPVNWD4+uWJGmajDpMB3AY8EJVfQ5Yn+SEMfVJkjRlRv3Z8auBTwBXttIbgD8Z\nV6ckSdNl1CujfwosBV4EqKqngMPH1SlJ0nQZNYxerqqi/YxEkjeNr0uSpGkzahjdnOSPgLck+S3g\nTvyhPUnSLBn1JyT+PXAL8GXgFOCqqvqPe2qXZGGSu5J8L8kjST7a6kckWZ1kbZLbk8wfanNlknVJ\nHk1y9lD9tCQPJ3ksybVD9YOT3NTa3JvkuKHnlrXt1ya5aJRjlSTtf3sMoyRzktxdVXdU1cer6t9U\n1R0j7n8z8DtV9U7gHwGXJ3k7cAVwZ1WdAtxFmxiR5FQGdwN/B3AucF2StH19Hri0qk4GTk6ypNUv\nBZ6tqpOAa4HPtn0dAVzF4DZGZwBXD4eeJGly7DGMqmoL8Oq+vJFX1caqeqgt/wh4lMGthc5j25dm\nVzD4Ei0MJkncVFWbq+pxBve+W5TkaODwqrq/bXfjUJvhfd0CnNmWlwCrq2pTVT0PrAbO2dtjkCSN\n36h3YPgR8EiSO2gz6gCq6qOjvlCStwLvBr4NLKiqmbaPjUmOapsdA9w71GxDq20G1g/V17f61jZP\ntn1tSbIpyZHD9R32JUmaMKOG0VfaY58keTODq5aPVdWPktQOm+y4/lpkz5tIkibJbsMoyXFV9YOq\n2uf70CWZyyCIVlbVra08k2RBVc20IbinW30DcOxQ84Wttqv6cJunkswB5lXVs0k2AIt3aHP3zvq4\nZcvLwPK2tniHZpIkuKc9xmNPnxl9betCki/v42v8MbCm3UZoq1XAxW15GXDrUP3CNkPuBOBE4L6q\n2ghsSrKoTWi4aIc2y9ryBQwmRADcDpyVZH6bzHBWq/2MOXMOZhBGyzGIJGlnFrPtfXL5rO99T8N0\nw0Neb9vbnSd5D/AbDD5vepDBcNwngd9j8N2lS4AnGMygo6rWJLkZWAO8AlzWvmwLcDlwA3AIcFtV\nfaPVrwdWJlkHPANc2Pb1XJJPAQ+0172mTWSQJE2YPYVR7WJ5JFX1LWDOLp5+3y7afBr49E7q3wHe\ntZP6S7Qw28lzNzAIMEnSBNtTGP1ikhcYXCEd2pZp61VV88baO0nSVNhtGFXVrq5qJEmaNXvze0aS\nJI2FYSRJ6s4wkiR1ZxhJkrozjCRJ3RlGkqTuDCNJUneGkSSpO8NIktSdYSRJ6s4wkiR1ZxhJkroz\njCRJ3RlGkqTuDCNJUneGkSSpO8NIktSdYSRJ6s4wkiR1ZxhJkrozjCRJ3aWqevehqyQ1d+5hbN78\nYtd+zJu3lIMOup/nn9/YtR8LFhzPxo2Pd+2DpN1LAvR+7w5Vldna29zZ2pFeu0EQ9f0Dm5mZtb8t\nSRqZw3SSpO4MI0lSd4aRJKk7w0iS1J1hJEnqzjCSJHVnGEmSujOMJEndGUaSpO4MI0lSd4aRJKm7\nsYZRkuuTzCR5eKh2RJLVSdYmuT3J/KHnrkyyLsmjSc4eqp+W5OEkjyW5dqh+cJKbWpt7kxw39Nyy\ntv3aJBeN8zglSa/NuK+Mvggs2aF2BXBnVZ0C3AVcCZDkVOBDwDuAc4HrMrg1LcDngUur6mTg5CRb\n93kp8GxVnQRcC3y27esI4CrgdOAM4Orh0JMkTZaxhlFV/QXw3A7l84AVbXkFcH5bXgrcVFWbq+px\nYB2wKMnRwOFVdX/b7sahNsP7ugU4sy0vAVZX1aaqeh5YDZwzawcmSZpVPT4zOqqqZgCqaiNwVKsf\nAzw5tN2GVjsGWD9UX99q27Wpqi3ApiRH7mZfkqQJNAm/ZzSbP+CzTz/Gs2XLy8Dytra4PSRJ29zT\nHuPRI4xmkiyoqpk2BPd0q28Ajh3abmGr7ao+3OapJHOAeVX1bJINbJ8oC4G7d9WhOXMOZvPm5ft+\nRJJ0wFvM9m+r18zq3vfHMF3Y/oplFXBxW14G3DpUv7DNkDsBOBG4rw3lbUqyqE1ouGiHNsva8gUM\nJkQA3A6clWR+m8xwVqtJkibQWK+MkvwZgyj9O0l+AFwNfAb4UpJLgCcYzKCjqtYkuRlYA7wCXFZV\nW4fwLgduAA4Bbquqb7T69cDKJOuAZ4AL276eS/Ip4AEGw4DXtIkMkqQJlG3v99MpSc2dexibN7/Y\ntR/z5i3lhRe+zux+hLYvwrT/TUiTbjBI1Pu/01BV+/Q5/c54BwZJUneTMJtOE+WNbPuucR8LFhzP\nxo2Pd+2DpP3LMNIOXqL35f/MTN8wlLT/OUwnSerOMJIkdWcYSZK6M4wkSd0ZRpKk7gwjSVJ3hpEk\nqTvDSJLUnV961QTqfxcI8E4Q0v5kGGkC9b8LBHgnCGl/cphOktSdYSRJ6s4wkiR1ZxhJkrpzAoO0\nS/1n9TmjT9PCMJJ2qf+sPmf0aVo4TCdJ6s4wkiR15zCdNNH6f24Ffnal8TOMpInW/3Mr8LMrjZ/D\ndJKk7rwykjSC/sOFDhUe2AwjSSPoP1zoUOGBzTCS9Drh1dmBzDCS9DoxCVdnh3QPxAOVYSRJI+sf\niAMHXiA6m06S1J1hJEnqzjCSJHVnGEmSujOMJEndGUaSpO4O+DBKck6Sv07yWJJP9O6PJOlnHdBh\nlOQg4D8BS4B3Ar+e5O19ezXJ7undgQlyT+8OTJB7endAU+CADiNgEbCuqp6oqleAm4DzOvdpgt3T\nuwMT5J7eHZgg9/TugKbAgR5GxwBPDq2vbzVJ0gTxdkBA1SvMm/fBrn14+eX7ur6+JPV0oIfRBuC4\nofWFrbadLVte4YUX/vt+69TuTcI9p+zDNpPQj0noA0xGP+zDNpPSj9mRqkm46d94JJkDrAV+Dfgh\ncB/w61X1aNeOSZK2c0BfGVXVliT/CljN4POx6w0iSZo8B/SVkSTp9eFAn023W9P2hdgkC5PcleR7\nSR5J8tFWPyLJ6iRrk9yeZP5QmyuTrEvyaJKz+/V+9iU5KMl3k6xq61N5HgCSzE/ypXZ830tyxjSe\njyS/neSvkjyc5E+THDxN5yHJ9Ulmkjw8VNvr409yWjuHjyW5dqQXr6qpfDAI4v8DHA+8AXgIeHvv\nfo35mI8G3t2W38zg87S3A78H/NtW/wTwmbZ8KvAgg+Hct7bzld7HMYvn47eBPwFWtfWpPA/tGG8A\nPtKW5wLzp+18AL8AfB84uK3/ObBsms4D8E+AdwMPD9X2+viBvwROb8u3AUv29NrTfGU0dV+IraqN\nVfVQW/4R8CiDGYbnASvaZiuA89vyUuCmqtpcVY8D6xict9e9JAuB9wNfGCpP3XkASDIPeG9VfRGg\nHecmpvN8zAHelGQucCiD2bdTcx6q6i+A53Yo79XxJzkaOLyq7m/b3TjUZpemOYym+guxSd7K4F9A\n3wYWVNUMDAILOKpttuM52sCBc47+APg42/+G9DSeB4ATgP+b5Itt2PK/JDmMKTsfVfUU8PvADxgc\n06aqupMpOw87cdReHv8xDN5PtxrpvXWaw2hqJXkzcAvwsXaFtOMslgN6VkuSDwAz7Spxd1/WOKDP\nw5C5wGnAf66q04AXgSuYvr+LtzC4CjiewZDdm5L8BlN2HkYwluOf5jAa6QuxB5o2/HALsLKqbm3l\nmSQL2vNHA0+3+gbg2KHmB8o5eg+wNMn3gf8GnJlkJbBxys7DVuuBJ6vqgbb+ZQbhNG1/F+8Dvl9V\nz1bVFuCrwD9m+s7Djvb2+PfpvExzGN0PnJjk+CQHAxcCqzr3aX/4Y2BNVX1uqLYKuLgtLwNuHapf\n2GYUnQCcyOCLw69rVfXJqjquqt7G4P/3u6rqw8DXmaLzsFUbgnkyycmt9GvA95iyvwsGw3P/MMkh\nScLgPKxh+s5D2H7EYK+Ovw3lbUqyqJ3Hi4ba7Frv2RudZ46cw2BG2Trgit792Q/H+x5gC4OZgw8C\n323n4EjgznYuVgNvGWpzJYNZMo8CZ/c+hjGck19h22y6aT4Pv8jgH2gPAV9hMJtu6s4HcHU7pocZ\nfFj/hmk6D8CfAU8BLzEI548AR+zt8QO/DDzS3ls/N8pr+6VXSVJ30zxMJ0maEIaRJKk7w0iS1J1h\nJEnqzjCSJHVnGEmSujOMJEndGUaSpO7+P4UePsTxCSasAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x2acb3be4a6d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "loci_df.QUAL.plot(kind=\"hist\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.Fis == -9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "56817"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL >= 10]) - len(loci_df[loci_df.QUAL >= 20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(122399, 65582)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL < 20]), len(loci_df[loci_df.QUAL < 10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(16515, 28411, 122951)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.Fis >= 0.5]), len(loci_df[loci_df.Fis <= -0.5]), len(loci_df[loci_df.MAF < 0.01])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def filter_snps(df, imputed=False):\n",
    "    if imputed:\n",
    "        return df[(df.MAF >= 0.01) & \n",
    "                  (df.Fis < 0.5) & \n",
    "                  (df.Fis > -0.5)]\n",
    "    else:\n",
    "        return df[(df.SUM_DEPTH >= 100) & \n",
    "                  (df.SUM_DEPTH < 1500) & \n",
    "                  (df.QUAL >= 20) & \n",
    "                  (df.MAF >= 0.01) & \n",
    "                  (df.Fis < 0.5) & \n",
    "                  (df.Fis > -0.5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(778406, 25)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_stage1 = filter_snps(loci_df)\n",
    "loci_stage1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUM_DEPTH</th>\n",
       "      <th>SUMSQ_DEPTH</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "      <th>VAR_DEPTH</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPE_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>N_CHR</th>\n",
       "      <th>A1_FREQ</th>\n",
       "      <th>A2_FREQ</th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "      <th>E(HOM1/HET/HOM2)</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "      <th>A1_allele</th>\n",
       "      <th>A2_allele</th>\n",
       "      <th>A1_freq</th>\n",
       "      <th>A2_freq</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Fis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE_73_length_45_cov_1.955556-63</th>\n",
       "      <td>1063</td>\n",
       "      <td>9127</td>\n",
       "      <td>4.35656</td>\n",
       "      <td>18.5020</td>\n",
       "      <td>334</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>96</td>\n",
       "      <td>0.196721</td>\n",
       "      <td>2</td>\n",
       "      <td>392</td>\n",
       "      <td>T:0.168367</td>\n",
       "      <td>C:0.831633</td>\n",
       "      <td>17/32/147</td>\n",
       "      <td>5.56/54.89/135.56</td>\n",
       "      <td>34.080940</td>\n",
       "      <td>1.870726e-07</td>\n",
       "      <td>1.870726e-07</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.168367</td>\n",
       "      <td>0.831633</td>\n",
       "      <td>0.168367</td>\n",
       "      <td>0.418479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_162_length_98_cov_5.010204-17</th>\n",
       "      <td>678</td>\n",
       "      <td>6288</td>\n",
       "      <td>2.77869</td>\n",
       "      <td>18.1237</td>\n",
       "      <td>371</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>216</td>\n",
       "      <td>0.442623</td>\n",
       "      <td>2</td>\n",
       "      <td>272</td>\n",
       "      <td>G:0.944853</td>\n",
       "      <td>A:0.0551471</td>\n",
       "      <td>121/15/0</td>\n",
       "      <td>121.41/14.17/0.41</td>\n",
       "      <td>0.463292</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.664991</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.944853</td>\n",
       "      <td>0.055147</td>\n",
       "      <td>0.055147</td>\n",
       "      <td>-0.054475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_162_length_98_cov_5.010204-27</th>\n",
       "      <td>732</td>\n",
       "      <td>7434</td>\n",
       "      <td>3.00000</td>\n",
       "      <td>21.5556</td>\n",
       "      <td>187</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>214</td>\n",
       "      <td>0.438525</td>\n",
       "      <td>2</td>\n",
       "      <td>274</td>\n",
       "      <td>C:0.967153</td>\n",
       "      <td>T:0.0328467</td>\n",
       "      <td>128/9/0</td>\n",
       "      <td>128.15/8.70/0.15</td>\n",
       "      <td>0.158021</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.873178</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.967153</td>\n",
       "      <td>0.032847</td>\n",
       "      <td>0.032847</td>\n",
       "      <td>-0.030189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_162_length_98_cov_5.010204-28</th>\n",
       "      <td>742</td>\n",
       "      <td>7634</td>\n",
       "      <td>3.04098</td>\n",
       "      <td>22.1300</td>\n",
       "      <td>181</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>214</td>\n",
       "      <td>0.438525</td>\n",
       "      <td>2</td>\n",
       "      <td>274</td>\n",
       "      <td>G:0.970803</td>\n",
       "      <td>A:0.0291971</td>\n",
       "      <td>129/8/0</td>\n",
       "      <td>129.12/7.77/0.12</td>\n",
       "      <td>0.123919</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.900253</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.970803</td>\n",
       "      <td>0.029197</td>\n",
       "      <td>0.029197</td>\n",
       "      <td>-0.026316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_162_length_98_cov_5.010204-31</th>\n",
       "      <td>744</td>\n",
       "      <td>7684</td>\n",
       "      <td>3.04918</td>\n",
       "      <td>22.2856</td>\n",
       "      <td>179</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>214</td>\n",
       "      <td>0.438525</td>\n",
       "      <td>2</td>\n",
       "      <td>274</td>\n",
       "      <td>C:0.956204</td>\n",
       "      <td>A:0.0437956</td>\n",
       "      <td>125/12/0</td>\n",
       "      <td>125.26/11.47/0.26</td>\n",
       "      <td>0.287396</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.777627</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.956204</td>\n",
       "      <td>0.043796</td>\n",
       "      <td>0.043796</td>\n",
       "      <td>-0.041985</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    SUM_DEPTH  SUMSQ_DEPTH  MEAN_DEPTH  \\\n",
       "NODE_73_length_45_cov_1.955556-63        1063         9127     4.35656   \n",
       "NODE_162_length_98_cov_5.010204-17        678         6288     2.77869   \n",
       "NODE_162_length_98_cov_5.010204-27        732         7434     3.00000   \n",
       "NODE_162_length_98_cov_5.010204-28        742         7634     3.04098   \n",
       "NODE_162_length_98_cov_5.010204-31        744         7684     3.04918   \n",
       "\n",
       "                                    VAR_DEPTH  QUAL  N_DATA  \\\n",
       "NODE_73_length_45_cov_1.955556-63     18.5020   334     488   \n",
       "NODE_162_length_98_cov_5.010204-17    18.1237   371     488   \n",
       "NODE_162_length_98_cov_5.010204-27    21.5556   187     488   \n",
       "NODE_162_length_98_cov_5.010204-28    22.1300   181     488   \n",
       "NODE_162_length_98_cov_5.010204-31    22.2856   179     488   \n",
       "\n",
       "                                    N_GENOTYPE_FILTERED  N_MISS    F_MISS  \\\n",
       "NODE_73_length_45_cov_1.955556-63                     0      96  0.196721   \n",
       "NODE_162_length_98_cov_5.010204-17                    0     216  0.442623   \n",
       "NODE_162_length_98_cov_5.010204-27                    0     214  0.438525   \n",
       "NODE_162_length_98_cov_5.010204-28                    0     214  0.438525   \n",
       "NODE_162_length_98_cov_5.010204-31                    0     214  0.438525   \n",
       "\n",
       "                                   N_ALLELES N_CHR     A1_FREQ      A2_FREQ  \\\n",
       "NODE_73_length_45_cov_1.955556-63          2   392  T:0.168367   C:0.831633   \n",
       "NODE_162_length_98_cov_5.010204-17         2   272  G:0.944853  A:0.0551471   \n",
       "NODE_162_length_98_cov_5.010204-27         2   274  C:0.967153  T:0.0328467   \n",
       "NODE_162_length_98_cov_5.010204-28         2   274  G:0.970803  A:0.0291971   \n",
       "NODE_162_length_98_cov_5.010204-31         2   274  C:0.956204  A:0.0437956   \n",
       "\n",
       "                                   OBS(HOM1/HET/HOM2)   E(HOM1/HET/HOM2)  \\\n",
       "NODE_73_length_45_cov_1.955556-63           17/32/147  5.56/54.89/135.56   \n",
       "NODE_162_length_98_cov_5.010204-17           121/15/0  121.41/14.17/0.41   \n",
       "NODE_162_length_98_cov_5.010204-27            128/9/0   128.15/8.70/0.15   \n",
       "NODE_162_length_98_cov_5.010204-28            129/8/0   129.12/7.77/0.12   \n",
       "NODE_162_length_98_cov_5.010204-31           125/12/0  125.26/11.47/0.26   \n",
       "\n",
       "                                    ChiSq_HWE         P_HWE  P_HET_DEFICIT  \\\n",
       "NODE_73_length_45_cov_1.955556-63   34.080940  1.870726e-07   1.870726e-07   \n",
       "NODE_162_length_98_cov_5.010204-17   0.463292  1.000000e+00   1.000000e+00   \n",
       "NODE_162_length_98_cov_5.010204-27   0.158021  1.000000e+00   1.000000e+00   \n",
       "NODE_162_length_98_cov_5.010204-28   0.123919  1.000000e+00   1.000000e+00   \n",
       "NODE_162_length_98_cov_5.010204-31   0.287396  1.000000e+00   1.000000e+00   \n",
       "\n",
       "                                    P_HET_EXCESS A1_allele A2_allele  \\\n",
       "NODE_73_length_45_cov_1.955556-63       1.000000         T         C   \n",
       "NODE_162_length_98_cov_5.010204-17      0.664991         G         A   \n",
       "NODE_162_length_98_cov_5.010204-27      0.873178         C         T   \n",
       "NODE_162_length_98_cov_5.010204-28      0.900253         G         A   \n",
       "NODE_162_length_98_cov_5.010204-31      0.777627         C         A   \n",
       "\n",
       "                                     A1_freq   A2_freq       MAF       Fis  \n",
       "NODE_73_length_45_cov_1.955556-63   0.168367  0.831633  0.168367  0.418479  \n",
       "NODE_162_length_98_cov_5.010204-17  0.944853  0.055147  0.055147 -0.054475  \n",
       "NODE_162_length_98_cov_5.010204-27  0.967153  0.032847  0.032847 -0.030189  \n",
       "NODE_162_length_98_cov_5.010204-28  0.970803  0.029197  0.029197 -0.026316  \n",
       "NODE_162_length_98_cov_5.010204-31  0.956204  0.043796  0.043796 -0.041985  "
      ]
     },
     "execution_count": 455,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_stage1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1029063, 25)"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beagle_stage1 = filter_snps(loci_df_beagle, imputed=True)\n",
    "beagle_stage1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with open(os.path.join(snp_dir, \"stage1_positions.txt\"), \"w\") as o:\n",
    "    for elem in loci_stage1.index:\n",
    "        o.write(\"%s\\n\" % \"\\t\".join(elem.split(\"-\")))\n",
    "        \n",
    "with open(os.path.join(beagle_dir, \"stage1_positions.txt\"), \"w\") as o:\n",
    "    for elem in beagle_stage1.index:\n",
    "        o.write(\"%s\\n\" % \"\\t\".join(elem.split(\"-\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /home/lindb/wbp/concatenated/snps/good_snps\n",
      "\t--positions /home/lindb/wbp/concatenated/snps/stage1_positions.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 778406 out of a possible 1300961 Sites\n",
      "Run Time = 1898.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/good_snps\n",
      "\t--positions /home/lindb/wbp/concatenated/snps/beagle40/stage1_positions.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 1029063 out of a possible 1300961 Sites\n",
      "Run Time = 680.00 seconds\n"
     ]
    }
   ],
   "source": [
    "for d, vcf_gz in zip([snp_dir, beagle_dir], [vcf_filtered_gz, beagle_vcf_gz]):\n",
    "    !$vcftools --gzvcf $vcf_gz \\\n",
    "    --remove-indels  \\\n",
    "    --remove-filtered-all \\\n",
    "    --recode \\\n",
    "    --recode-INFO-all \\\n",
    "    --positions {os.path.join(d, \"stage1_positions.txt\")} \\\n",
    "    --out {os.path.join(d, \"good_snps\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n"
     ]
    }
   ],
   "source": [
    "for d in [snp_dir, beagle_dir]:\n",
    "    snps = os.path.join(d, \"good_snps.recode.vcf\")\n",
    "    snps_gz = snps + \".gz\"\n",
    "    !$bgzip -c {snps} > {snps_gz}\n",
    "    !$tabix {snps_gz}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_intersection(imp, ni):\n",
    "    return set.intersection(set(ni.index), set(imp.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "isect = get_intersection(beagle_stage1, loci_stage1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "isect = sorted(isect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(778406, 1029063, 713745)"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_stage1.index), len(beagle_stage1.index), len(isect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for d in [snp_dir, beagle_dir]:\n",
    "    with open(os.path.join(d, \"isect_positions.txt\"), \"w\") as o:\n",
    "        for elem in isect:\n",
    "            o.write(\"%s\\n\" % \"\\t\".join(elem.split(\"-\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /home/lindb/wbp/concatenated/snps/isect_snps\n",
      "\t--positions /home/lindb/wbp/concatenated/snps/isect_positions.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 713745 out of a possible 1300961 Sites\n",
      "Run Time = 1725.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/isect_snps\n",
      "\t--positions /home/lindb/wbp/concatenated/snps/beagle40/isect_positions.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 713745 out of a possible 1300961 Sites\n",
      "Run Time = 486.00 seconds\n"
     ]
    }
   ],
   "source": [
    "for d, vcf_gz in zip([snp_dir, beagle_dir], [vcf_filtered_gz, beagle_vcf_gz]):\n",
    "    !$vcftools --gzvcf $vcf_gz \\\n",
    "    --remove-indels  \\\n",
    "    --remove-filtered-all \\\n",
    "    --recode \\\n",
    "    --recode-INFO-all \\\n",
    "    --positions {os.path.join(d, \"isect_positions.txt\")} \\\n",
    "    --out {os.path.join(d, \"isect_snps\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcfsort = '/home/lindb/g/src/vcftools_0.1.13/perl/vcf-sort'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "sort -k1,1d -k2,2n\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "sort -k1,1d -k2,2n\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n"
     ]
    }
   ],
   "source": [
    "for d in [snp_dir, beagle_dir]:\n",
    "    snps = os.path.join(d, \"isect_snps.recode.vcf\")\n",
    "    snps_gz = snps + \".gz\"\n",
    "    !$bgzip -c {snps} > {snps_gz}\n",
    "    !$tabix {snps_gz}\n",
    "    \n",
    "    srted = snps_gz + \"_sorted.vcf\"\n",
    "    srted_gz = srted + \".gz\"\n",
    "    !$vcfsort {snps_gz} > {srted}\n",
    "    !$bgzip -c {srted} > {srted_gz}\n",
    "    !$tabix {srted_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# choose one snp per contig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 389,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/home/lindb/wbp/concatenated/snps/',\n",
       " '/home/lindb/wbp/concatenated/snps/beagle40')"
      ]
     },
     "execution_count": 389,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_dir,beagle_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 654,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#because I want to go to bed and this will take a while with 713K recs in vcf:\n",
    "pytext = '''from __future__ import division\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os, sys,time\n",
    "import vcf\n",
    "from collections import OrderedDict, Counter\n",
    "import random\n",
    "\n",
    "VCFmissing = '/home/lindb/wbp/concatenated/snps/isect_snps.recode.vcf'\n",
    "missing_reader = vcf.Reader(open(VCFmissing),'r')\n",
    "\n",
    "filE1 = '/home/lindb/wbp/concatenated/snps/update.txt'\n",
    "if os.path.exists(filE1):\n",
    "    os.remove(filE1)\n",
    "with open(filE1,'w') as o:\n",
    "    text = 'starting recs \\\\n'\n",
    "    o.write(\"%s\" % text)\n",
    "\n",
    "\n",
    "locDict = OrderedDict()\n",
    "loccount = 0\n",
    "for rec in missing_reader:\n",
    "    contig = rec.CHROM\n",
    "    locus = rec.POS\n",
    "    if contig not in locDict.keys():\n",
    "        locDict[contig] = Counter()\n",
    "\n",
    "    womp = 0\n",
    "    for samp in rec.samples:\n",
    "        gt = samp['GT']\n",
    "\n",
    "        if '.' in gt:\n",
    "            try:\n",
    "                assert gt == './.'\n",
    "                womp += 1 #womp womp\n",
    "            except Exception as E:\n",
    "                print gt,E\n",
    "        else: #add count for a sample without missing data\n",
    "            locDict[contig][str(locus)] += 1\n",
    "\n",
    "    assert womp + locDict[contig][str(locus)] == len(rec.samples)\n",
    "    loccount += 1\n",
    "    if loccount % 1000 == 0:\n",
    "        with open(filE1,'a') as o:\n",
    "            o.write(\"%s\\\\n\" % str(loccount))\n",
    "    #break\n",
    "\n",
    "with open(filE1,'a') as o:\n",
    "    o.write(\"choosing snps \\\\n\")\n",
    "\n",
    "keep = OrderedDict()\n",
    "contcount = 0\n",
    "for contig in locDict.keys():\n",
    "    #print contig\n",
    "    #print len(locDict[contig].keys())\n",
    "\n",
    "    #find SNP with least missing data\n",
    "    mx = 0\n",
    "    mxloc = ['Hello How Are You?']\n",
    "    for locus in locDict[contig].keys():\n",
    "        #print locus,locDict[contig][locus]\n",
    "\n",
    "        if locDict[contig][locus] > mx: #if > max count\n",
    "            mx = locDict[contig][locus]\n",
    "            mxloc = [str(locus)]\n",
    "            #print mxloc\n",
    "        elif locDict[contig][locus] == mx: #if = max count\n",
    "            mxloc.append(str(locus))\n",
    "            #print mxloc\n",
    "\n",
    "    assert mxloc != ['Hello How Are You?']\n",
    "    if len(mxloc) > 1:\n",
    "        x = random.random()\n",
    "\n",
    "        bins = []\n",
    "        for i in range(len(mxloc)):\n",
    "            if i == 0:\n",
    "                bins.append(0)\n",
    "            else:\n",
    "                bins.append(i/len(mxloc))\n",
    "        bins.append(1)\n",
    "\n",
    "        for b in range(len(bins)):\n",
    "            if (x > bins[b]) and (x <= bins[b+1]):\n",
    "                lowcuss = mxloc[b]\n",
    "                #print \"b=\",b\n",
    "                try:\n",
    "                    print snp\n",
    "                    sys.exit(\"snp already exists. WTF?\")\n",
    "                except:\n",
    "                    pass\n",
    "                snp = \"-\".join([contig,str(lowcuss)])\n",
    "                break #keep\n",
    "    else:\n",
    "        snp = \"-\".join([contig,str(mxloc[0])])\n",
    "    #print snp\n",
    "\n",
    "    assert contig not in keep.keys()\n",
    "    keep[contig] = snp\n",
    "    \n",
    "    #delete just in case\n",
    "    del snp, locus\n",
    "    try:\n",
    "        del lowcuss\n",
    "    except:\n",
    "        pass\n",
    "    contcount += 1\n",
    "    if contcount % 1000 == 0:\n",
    "        with open(filE1,'a') as o:\n",
    "            o.write(\"%s\" % str(contcount))\n",
    "\n",
    "with open(filE1,'a') as o:\n",
    "    o.write(\"writing file\\\\n\")\n",
    "\n",
    "filE = '/home/lindb/wbp/concatenated/snps/one_snp_per_contig.txt'\n",
    "if os.path.exists(filE):\n",
    "    os.remove(filE)\n",
    "with open(filE,'w') as o:\n",
    "    for k in keep.keys():\n",
    "        text = '\\\\t'.join(keep[k].split(\"-\")) + '\\\\n'\n",
    "        o.write(\"%s\\\\n\" % text)\n",
    "\n",
    "'''\n",
    "filE = '/home/lindb/wbp/concatenated/snps/get_one_snp_per_contig.py'\n",
    "with open(filE,'w') as o:\n",
    "    o.write(\"%s\" % pytext)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 641,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "shtext = '''#!/bin/bash\n",
    "#$ -N contigs\n",
    "#$ -V\n",
    "#$ -j y\n",
    "#$ -cwd\n",
    "\n",
    "cd /home/lindb/wbp/concatenated/snps/\n",
    "python get_one_snp_per_contig.py\n",
    "\n",
    "'''\n",
    "filE = '/home/lindb/wbp/concatenated/snps/run_get_one_snp_per_contig.sh'\n",
    "with open(filE,'w') as o:\n",
    "    o.write(\"%s\" % shtext)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 647,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/gpfs_fs/home/lindb/wbp/concatenated/snps\n"
     ]
    }
   ],
   "source": [
    "cd ~/wbp/concatenated/snps/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 649,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\r\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\r\n",
      "Your job 1416914 (\"get_one_snp_per_contig.py\") has been submitted\r\n"
     ]
    }
   ],
   "source": [
    "!qsub $filE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 659,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 661,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "shutil.copy('/home/lindb/wbp/concatenated/snps/one_snp_per_contig.txt',\n",
    "            '/home/lindb/wbp/concatenated/snps/beagle40/one_snp_per_contig.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 662,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NODE_73_length_45_cov_1.955556</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NODE_162_length_98_cov_5.010204</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NODE_411_length_91_cov_2.098901</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NODE_550_length_45_cov_3.133333</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NODE_753_length_64_cov_3.093750</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 0   1\n",
       "0   NODE_73_length_45_cov_1.955556  63\n",
       "1  NODE_162_length_98_cov_5.010204  27\n",
       "2  NODE_411_length_91_cov_2.098901  18\n",
       "3  NODE_550_length_45_cov_3.133333  44\n",
       "4  NODE_753_length_64_cov_3.093750  54"
      ]
     },
     "execution_count": 662,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('/home/lindb/wbp/concatenated/snps/one_snp_per_contig.txt',header=None,sep=\"\\t\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 667,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(116231, 2)"
      ]
     },
     "execution_count": 667,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 665,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "116231"
      ]
     },
     "execution_count": 665,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.unique(df[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 666,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "116231"
      ]
     },
     "execution_count": 666,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 672,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/home/lindb/wbp/concatenated/snps/samtools_1.3.vcf.gz.recode.vcf.gz',\n",
       " '/home/lindb/wbp/concatenated/snps/beagle40/beagle40.vcf.gz']"
      ]
     },
     "execution_count": 672,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[vcf_filtered_gz, beagle_vcf_gz]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 673,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcf_filtered_gz_safe = '/home/lindb/wbp/concatenated/snps/onepcontig/samtools_1.3.vcf.gz.recode.vcf.gz'\n",
    "beagle_vcf_gz_safe = '/home/lindb/wbp/concatenated/snps/onepcontig/beagle40.vcf.gz'\n",
    "assert os.path.exists(vcf_filtered_gz_safe)\n",
    "assert os.path.exists(beagle_vcf_gz_safe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 674,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/onepcontig/samtools_1.3.vcf.gz.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /home/lindb/wbp/concatenated/snps/isect_one_per_contig\n",
      "\t--positions /home/lindb/wbp/concatenated/snps/one_snp_per_contig.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 116231 out of a possible 1300961 Sites\n",
      "Run Time = 397.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/onepcontig/beagle40.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig\n",
      "\t--positions /home/lindb/wbp/concatenated/snps/beagle40/one_snp_per_contig.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 116231 out of a possible 1300961 Sites\n",
      "Run Time = 154.00 seconds\n"
     ]
    }
   ],
   "source": [
    "#write files\n",
    "for d, vcf_gz in zip([snp_dir, beagle_dir], [vcf_filtered_gz_safe, beagle_vcf_gz_safe]):\n",
    "    !$vcftools --gzvcf $vcf_gz \\\n",
    "    --remove-indels  \\\n",
    "    --remove-filtered-all \\\n",
    "    --recode \\\n",
    "    --recode-INFO-all \\\n",
    "    --positions {os.path.join(d, \"one_snp_per_contig.txt\")} \\\n",
    "    --out {os.path.join(d, \"isect_one_per_contig\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 675,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n"
     ]
    }
   ],
   "source": [
    "#zip files\n",
    "for d in [snp_dir, beagle_dir]:\n",
    "    snps = os.path.join(d, \"isect_one_per_contig.recode.vcf\")\n",
    "    snps_gz = snps + \".gz\"\n",
    "    !$bgzip -c {snps} > {snps_gz}\n",
    "    !$tabix {snps_gz}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 683,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 683,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.path.exists(snps_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 686,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/lindb/g/src/vcftools_0.1.13/perl/vcf-sort'"
      ]
     },
     "execution_count": 686,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcfsort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 687,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "sort -k1,1d -k2,2n\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "sort -k1,1d -k2,2n\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n"
     ]
    }
   ],
   "source": [
    "for d in [snp_dir, beagle_dir]:\n",
    "    snps = os.path.join(d, \"isect_one_per_contig.recode.vcf\")\n",
    "    snps_gz = snps + \".gz\"\n",
    "    srted = snps_gz + \"_sorted.vcf\"\n",
    "    srted_gz = srted + \".gz\"\n",
    "\n",
    "    !$vcfsort {snps_gz} > {srted}\n",
    "    !$bgzip -c {srted} > {srted_gz}\n",
    "    !$tabix {srted_gz}    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 695,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz\n",
      "\t--chrom-map /home/lindb/wbp/concatenated/snps/chrom_map.txt\n",
      "\t--out /home/lindb/wbp/concatenated/snps/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz\n",
      "\t--plink\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Writing PLINK PED and MAP files ... \n",
      "\tRead 178750 chromosome mapping file entries.\n",
      "Done.\n",
      "After filtering, kept 116231 out of a possible 116231 Sites\n",
      "Run Time = 35.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "@----------------------------------------------------------@\n",
      "|        PLINK!       |     v1.07      |   10/Aug/2009     |\n",
      "|----------------------------------------------------------|\n",
      "|  (C) 2009 Shaun Purcell, GNU General Public License, v2  |\n",
      "|----------------------------------------------------------|\n",
      "|  For documentation, citation & bug-report instructions:  |\n",
      "|        http://pngu.mgh.harvard.edu/purcell/plink/        |\n",
      "@----------------------------------------------------------@\n",
      "\n",
      "Skipping web check... [ --noweb ] \n",
      "Writing this text to log file [ /home/lindb/wbp/concatenated/snps/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz_recodeA.log ]\n",
      "Analysis started: Wed Feb 17 14:01:35 2016\n",
      "\n",
      "Options in effect:\n",
      "\t--noweb\n",
      "\t--recodeA\n",
      "\t--tab\n",
      "\t--file /home/lindb/wbp/concatenated/snps/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz\n",
      "\t--out /home/lindb/wbp/concatenated/snps/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz_recodeA\n",
      "\n",
      "** For gPLINK compatibility, do not use '.' in --out **\n",
      "116231 (of 116231) markers to be included from [ /home/lindb/wbp/concatenated/snps/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz.map ]\n",
      "Warning, found 244 individuals with ambiguous sex codes\n",
      "Writing list of these individuals to [ /home/lindb/wbp/concatenated/snps/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz_recodeA.nosex ]\n",
      "244 individuals read from [ /home/lindb/wbp/concatenated/snps/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz.ped ] \n",
      "0 individuals with nonmissing phenotypes\n",
      "Assuming a disease phenotype (1=unaff, 2=aff, 0=miss)\n",
      "Missing phenotype value is also -9\n",
      "0 cases, 0 controls and 244 missing\n",
      "0 males, 0 females, and 244 of unspecified sex\n",
      "Before frequency and genotyping pruning, there are 116231 SNPs\n",
      "244 founders and 0 non-founders found\n",
      "28 heterozygous haploid genotypes; set to missing\n",
      "Writing list of heterozygous haploid genotypes to [ /home/lindb/wbp/concatenated/snps/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz_recodeA.hh ]\n",
      "Total genotyping rate in remaining individuals is 0.649729\n",
      "0 SNPs failed missingness test ( GENO > 1 )\n",
      "0 SNPs failed frequency test ( MAF < 0 )\n",
      "After frequency and genotyping pruning, there are 116231 SNPs\n",
      "After filtering, 0 cases, 0 controls and 244 missing\n",
      "After filtering, 0 males, 0 females, and 244 of unspecified sex\n",
      "Writing recoded file to [ /home/lindb/wbp/concatenated/snps/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz_recodeA.raw ] \n",
      "\n",
      "Analysis finished: Wed Feb 17 14:02:03 2016\n",
      "\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz\n",
      "\t--chrom-map /home/lindb/wbp/concatenated/snps/chrom_map.txt\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz\n",
      "\t--plink\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Writing PLINK PED and MAP files ... \n",
      "\tRead 178750 chromosome mapping file entries.\n",
      "Done.\n",
      "After filtering, kept 116231 out of a possible 116231 Sites\n",
      "Run Time = 26.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "@----------------------------------------------------------@\n",
      "|        PLINK!       |     v1.07      |   10/Aug/2009     |\n",
      "|----------------------------------------------------------|\n",
      "|  (C) 2009 Shaun Purcell, GNU General Public License, v2  |\n",
      "|----------------------------------------------------------|\n",
      "|  For documentation, citation & bug-report instructions:  |\n",
      "|        http://pngu.mgh.harvard.edu/purcell/plink/        |\n",
      "@----------------------------------------------------------@\n",
      "\n",
      "Skipping web check... [ --noweb ] \n",
      "Writing this text to log file [ /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz_recodeA.log ]\n",
      "Analysis started: Wed Feb 17 14:02:30 2016\n",
      "\n",
      "Options in effect:\n",
      "\t--noweb\n",
      "\t--recodeA\n",
      "\t--tab\n",
      "\t--file /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz_recodeA\n",
      "\n",
      "** For gPLINK compatibility, do not use '.' in --out **\n",
      "116231 (of 116231) markers to be included from [ /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz.map ]\n",
      "Warning, found 244 individuals with ambiguous sex codes\n",
      "Writing list of these individuals to [ /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz_recodeA.nosex ]\n",
      "244 individuals read from [ /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz.ped ] \n",
      "0 individuals with nonmissing phenotypes\n",
      "Assuming a disease phenotype (1=unaff, 2=aff, 0=miss)\n",
      "Missing phenotype value is also -9\n",
      "0 cases, 0 controls and 244 missing\n",
      "0 males, 0 females, and 244 of unspecified sex\n",
      "Before frequency and genotyping pruning, there are 116231 SNPs\n",
      "244 founders and 0 non-founders found\n",
      "28 heterozygous haploid genotypes; set to missing\n",
      "Writing list of heterozygous haploid genotypes to [ /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz_recodeA.hh ]\n",
      "Total genotyping rate in remaining individuals is 1\n",
      "0 SNPs failed missingness test ( GENO > 1 )\n",
      "0 SNPs failed frequency test ( MAF < 0 )\n",
      "After frequency and genotyping pruning, there are 116231 SNPs\n",
      "After filtering, 0 cases, 0 controls and 244 missing\n",
      "After filtering, 0 males, 0 females, and 244 of unspecified sex\n",
      "Writing recoded file to [ /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz_recodeA.raw ] \n",
      "\n",
      "Analysis finished: Wed Feb 17 14:02:59 2016\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for d in [snp_dir, beagle_dir]:\n",
    "    f = os.path.join(d, \"isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz\")\n",
    "    assert os.path.exists(f)\n",
    "    write_plink_files(f)\n",
    "    write_plink_recode(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 696,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz\n",
      "\t--012\n",
      "\t--out /home/lindb/wbp/concatenated/snps/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 116231 out of a possible 116231 Sites\n",
      "Run Time = 30.00 seconds\n",
      "/bin/sh: module: line 1: syntax error: unexpected end of file\n",
      "/bin/sh: error importing function definition for `BASH_FUNC_module'\n",
      "\n",
      "VCFtools - v0.1.13\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz\n",
      "\t--012\n",
      "\t--out /home/lindb/wbp/concatenated/snps/beagle40/isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.3\n",
      "Versions of zlib >= 1.2.4 will be *much* faster when reading zipped VCF files.\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 116231 out of a possible 116231 Sites\n",
      "Run Time = 20.00 seconds\n"
     ]
    }
   ],
   "source": [
    "for d in [snp_dir, beagle_dir]:\n",
    "    f = os.path.join(d, \"isect_one_per_contig.recode.vcf.gz_sorted.vcf.gz\")\n",
    "    assert os.path.exists(f)\n",
    "    !$vcftools --gzvcf {f} \\\n",
    "    --out {f} \\\n",
    "    --012"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 617,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record(CHROM=NODE_17604_length_98_cov_2.367347, POS=92, REF=G, ALT=[A])\n"
     ]
    }
   ],
   "source": [
    "VCFmissing = os.path.join(snp_dir,'isect_snps.recode.vcf')\n",
    "missing_reader = vcf.Reader(open(VCFmissing),'r')\n",
    "\n",
    "locDict = OrderedDict()\n",
    "for rec in missing_reader:\n",
    "    print rec\n",
    "    contig = rec.CHROM\n",
    "    locus = rec.POS\n",
    "    if contig not in locDict.keys():\n",
    "        locDict[contig] = Counter()\n",
    "\n",
    "    womp = 0\n",
    "    for i,samp in enumerate(rec.samples): \n",
    "        gt = samp['GT']\n",
    "\n",
    "        if '.' in gt:\n",
    "            try:\n",
    "                assert gt == './.'\n",
    "                womp += 1 #womp womp\n",
    "            except Exception as E:\n",
    "                print gt,E\n",
    "        else: #add count for a sample without missing data\n",
    "            locDict[contig][str(locus)] += 1\n",
    "\n",
    "    assert womp + locDict[contig][str(locus)] == len(rec.samples)\n",
    "    break\n",
    "\n",
    "keep = OrderedDict()\n",
    "for contig in locDict.keys():\n",
    "    #print contig\n",
    "    #print len(locDict[contig].keys())\n",
    "\n",
    "    #find SNP with least missing data\n",
    "    mx = 0\n",
    "    mxloc = ['Hello How Are You?']\n",
    "    for locus in locDict[contig].keys():\n",
    "        #print locus,locDict[contig][locus]\n",
    "\n",
    "        if locDict[contig][locus] > mx: #if > max count\n",
    "            mx = locDict[contig][locus]\n",
    "            mxloc = [str(locus)]\n",
    "            #print mxloc\n",
    "        elif locDict[contig][locus] == mx: #if = max count\n",
    "            mxloc.append(str(locus))\n",
    "            #print mxloc\n",
    "\n",
    "    assert mxloc != ['Hello How Are You?']\n",
    "    if len(mxloc) > 1:\n",
    "        x = random.random()\n",
    "\n",
    "        bins = []\n",
    "        for i in range(len(mxloc)):\n",
    "            if i == 0:\n",
    "                bins.append(0)\n",
    "            else:\n",
    "                bins.append(i/len(mxloc))\n",
    "        bins.append(1)\n",
    "\n",
    "        for b in range(len(bins)):\n",
    "            if (x > bins[b]) and (x <= bins[b+1]):\n",
    "                lowcuss = mxloc[b]\n",
    "                #print \"b=\",b\n",
    "                try:\n",
    "                    print snp\n",
    "                    sys.exit(\"snp already exists. WTF?\")\n",
    "                except:\n",
    "                    pass\n",
    "                snp = \"-\".join([contig,str(lowcuss)])\n",
    "                break #keep\n",
    "    else:\n",
    "        snp = \"-\".join([contig,str(mxloc[0])])\n",
    "    #print snp\n",
    "\n",
    "    assert contig not in keep.keys()\n",
    "    keep[contig] = snp\n",
    "    \n",
    "    #delete just in case\n",
    "    del snp, locus\n",
    "    try:\n",
    "        del lowcuss\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "filE = '/home/lindb/wbp/concatenated/snps/one_snp_per_contig.txt'\n",
    "if os.path.exists(filE):\n",
    "    os.remove(filE)\n",
    "with open(filE,'w') as o:\n",
    "    for k in keep.keys():\n",
    "        text = '\\t'.join(keep[k].split(\"-\")) + '\\n'\n",
    "        o.write(\"%s\" % text)\n",
    "        \n",
    "df = pd.read_csv(filE,sep=\"\\t\",header=None)\n",
    "df    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
